<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>A Swift, Versatile Phylogenomic and High-Throughput Sequencing Simulator • jackalope</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="A Swift, Versatile Phylogenomic and High-Throughput Sequencing Simulator">
<meta property="og:description" content="Simply and efficiently
    simulates (i) variants from reference genomes and (ii) reads from both Illumina 
    &lt;https://www.illumina.com/&gt;
    and Pacific Biosciences (PacBio) &lt;https://www.pacb.com/&gt; platforms. 
    It can either read reference genomes from FASTA files or simulate new ones.
    Genomic variants can be simulated using summary statistics, phylogenies, 
    Variant Call Format (VCF) files, and coalescent simulations—the latter of which
    can include selection, recombination, and demographic fluctuations.
    jackalope can simulate single, paired-end, or mate-pair Illumina reads, 
    as well as PacBio reads.
    These simulations include sequencing errors, mapping qualities, multiplexing,
    and optical/polymerase chain reaction (PCR) duplicates.
    Simulating Illumina sequencing is based on ART
    by Huang et al. (2012) &lt;doi:10.1093/bioinformatics/btr708&gt;.
    PacBio sequencing simulation is based on 
    SimLoRD  by Stöcker et al. (2016) &lt;doi:10.1093/bioinformatics/btw286&gt;.
    All outputs can be written to standard file formats.">
<meta property="og:image" content="/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">jackalope</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.1.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/jackalope-intro.html">Intro to jackalope</a>
    </li>
    <li>
      <a href="articles/sub-models.html">Models of nucleotide substitution</a>
    </li>
  </ul>
</li>
<li>
  <a href="news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/lucasnell/jackalope/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">


<div id="jackalope-" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#jackalope-" class="anchor"></a>jackalope <img src="reference/figures/logo.png" align="right" alt="" width="120">
</h1></div>
<div id="overview" class="section level2">
<h2 class="hasAnchor">
<a href="#overview" class="anchor"></a>Overview</h2>
<p>For studies using high-throughput sequencing (HTS) data, simulations can be vital for planning sampling design and testing bioinformatic tools. However, most HTS sequencing tools provide only very simple ways of adding deviations from a reference genome. For HTS studies that focus on patterns of genomic variation among individuals, populations, or species, having a tool that can simulate realistic patterns of molecular evolution and generate HTS data from those simulations would be quite useful.</p>
<p><code>jackalope</code> simply and efficiently simulates (i) haplotypes from reference genomes and (ii) reads from both Illumina and Pacific Biosciences (PacBio) platforms. It can either read reference genomes from FASTA files or simulate new ones. Variant haplotypes can be simulated using summary statistics, phylogenies, Variant Call Format (VCF) files, and coalescent simulations—the latter of which can include selection, recombination, and demographic fluctuations. <code>jackalope</code> can simulate single, paired-end, or mate-pair Illumina reads, as well as reads from Pacific Biosciences These simulations include sequencing errors, mapping qualities, multiplexing, and optical/PCR duplicates. All outputs can be written to standard file formats.</p>
</div>
<div id="installation" class="section level2">
<h2 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h2>
<div id="dependencies" class="section level3">
<h3 class="hasAnchor">
<a href="#dependencies" class="anchor"></a>Dependencies</h3>
<p>Before installing <code>jackalope</code>, you should update the packages <code>Rhtslib</code> and <code>zlibbioc</code>. Since both of these are on Bioconductor, you should update <code>BiocManager</code>, too.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">||</span>
    <span class="st">"BiocManager"</span> <span class="op">%in%</span> <span class="fu"><a href="https://rdrr.io/r/base/row.names.html">row.names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/update.packages.html">old.packages</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span>
  <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span>
<span class="op">}</span>
<span class="fu">BiocManager</span><span class="fu">::</span><span class="fu">install</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Rhtslib"</span>, <span class="st">"zlibbioc"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
</div>
<div id="stable-version" class="section level3">
<h3 class="hasAnchor">
<a href="#stable-version" class="anchor"></a>Stable version</h3>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># To install the latest stable version from CRAN:</span>
<span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"jackalope"</span><span class="op">)</span></code></pre></div>
</div>
<div id="development-version" class="section level3">
<h3 class="hasAnchor">
<a href="#development-version" class="anchor"></a>Development version</h3>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># install.packages("devtools")</span>
<span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html">install_github</a></span><span class="op">(</span><span class="st">"lucasnell/jackalope"</span><span class="op">)</span></code></pre></div>
</div>
<div id="enabling-openmp" class="section level3">
<h3 class="hasAnchor">
<a href="#enabling-openmp" class="anchor"></a>Enabling OpenMP</h3>
<p>To use multithreading in <code>jackalope</code>, you’ll need to compile it from source using the proper flags. If you’ve enabled OpenMP properly, running <code>jackalope:::using_openmp()</code> in R should return <code>TRUE</code>.</p>
<div id="windows-and-linux" class="section level4">
<h4 class="hasAnchor">
<a href="#windows-and-linux" class="anchor"></a>Windows and Linux</h4>
<p>The first step is to add the following to the <code>.R/Makevars</code> (<code>.R/Makevars.win</code> on Windows) file inside the home directory:</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1"></a><span class="ex">PKG_CXXFLAGS</span> += <span class="va">$(</span><span class="ex">SHLIB_OPENMP_CXXFLAGS</span><span class="va">)</span></span>
<span id="cb4-2"><a href="#cb4-2"></a><span class="ex">PKG_CFLAGS</span> += <span class="va">$(</span><span class="ex">SHLIB_OPENMP_CFLAGS</span><span class="va">)</span></span>
<span id="cb4-3"><a href="#cb4-3"></a><span class="ex">PKG_LIBS</span> += <span class="va">$(</span><span class="ex">SHLIB_OPENMP_CFLAGS</span><span class="va">)</span></span></code></pre></div>
<p>Then, you should be able to install <code>jackalope</code> by running the following in R:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"jackalope"</span>, type <span class="op">=</span> <span class="st">"source"</span><span class="op">)</span>
<span class="co">## Or, for development version:</span>
<span class="co"># remotes::install_github("lucasnell/jackalope")</span></code></pre></div>
</div>
<div id="macos-r-version--400" class="section level4">
<h4 class="hasAnchor">
<a href="#macos-r-version--400" class="anchor"></a>macOS, R version &gt;= 4.0.0</h4>
<p>Follow the directions here to install R compiler tools: <a href="https://thecoatlessprofessor.com/programming/cpp/r-compiler-tools-for-rcpp-on-macos/" class="uri">https://thecoatlessprofessor.com/programming/cpp/r-compiler-tools-for-rcpp-on-macos/</a>.</p>
<p>Check your version of <code>gcc</code> using <code>gcc --version</code> in the Terminal. Then, check the table at <a href="https://mac.r-project.org/openmp/" class="uri">https://mac.r-project.org/openmp/</a> to see which version of the runtime OpenMP downloads you need. For LLVM version 9.0.1, you run the following in the Terminal:</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1"></a><span class="bu">export</span> <span class="va">version=</span><span class="st">"9.0.1"</span></span>
<span id="cb6-2"><a href="#cb6-2"></a></span>
<span id="cb6-3"><a href="#cb6-3"></a><span class="ex">curl</span> -O https://mac.r-project.org/openmp/openmp-<span class="va">${version}</span>-darwin17-Release.tar.gz</span>
<span id="cb6-4"><a href="#cb6-4"></a><span class="fu">sudo</span> tar fvx openmp-<span class="va">${version}</span>-darwin17-Release.tar.gz -C /</span></code></pre></div>
<p>For the next step of actually installing <code>jackalope</code>, one option is to add the following to your <code>~/.R/Makevars</code> file:</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1"></a><span class="ex">CPPFLAGS</span> += -Xclang -fopenmp</span>
<span id="cb7-2"><a href="#cb7-2"></a><span class="ex">LDFLAGS</span> += -lomp</span></code></pre></div>
<p>… then installing <code>jackalope</code> by running <code><a href="https://rdrr.io/r/utils/install.packages.html">install.packages("jackalope", type = "source")</a></code> or <code><a href="https://remotes.r-lib.org/reference/install_github.html">remotes::install_github("lucasnell/jackalope")</a></code> in R.</p>
<p>This might not be desirable since it affects all package installations. An alternative method is to use the package <code>withr</code>:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">withr</span><span class="fu">::</span><span class="fu"><a href="https://withr.r-lib.org/reference/with_makevars.html">with_makevars</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span>CPPFLAGS <span class="op">=</span> <span class="st">"-Xclang -fopenmp"</span>, LDFLAGS <span class="op">=</span> <span class="st">"-lomp"</span><span class="op">)</span>, 
                     <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"jackalope"</span>, type <span class="op">=</span> <span class="st">"source"</span><span class="op">)</span>,
                     <span class="co">## For development version:</span>
                     <span class="co"># remotes::install_github("lucasnell/jackalope"),</span>
                     assignment <span class="op">=</span> <span class="st">"+="</span><span class="op">)</span></code></pre></div>
</div>
<div id="macos-r-version--34-and--400" class="section level4">
<h4 class="hasAnchor">
<a href="#macos-r-version--34-and--400" class="anchor"></a>macOS, R version &gt;= 3.4* and &lt; 4.0.0</h4>
<p>Add the following to the <code>.R/Makevars</code> file inside the home directory:</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb9-1"><a href="#cb9-1"></a><span class="ex">PKG_CXXFLAGS</span> += <span class="va">$(</span><span class="ex">SHLIB_OPENMP_CXXFLAGS</span><span class="va">)</span></span>
<span id="cb9-2"><a href="#cb9-2"></a><span class="ex">PKG_CFLAGS</span> += <span class="va">$(</span><span class="ex">SHLIB_OPENMP_CFLAGS</span><span class="va">)</span></span>
<span id="cb9-3"><a href="#cb9-3"></a><span class="ex">PKG_LIBS</span> += <span class="va">$(</span><span class="ex">SHLIB_OPENMP_CFLAGS</span><span class="va">)</span></span></code></pre></div>
<p>Next, go to <a href="https://cran.r-project.org/bin/macosx/tools/" class="uri">https://cran.r-project.org/bin/macosx/tools/</a> and download the newest versions of (1) the <code>clang</code> compiler (version 8 at the time of writing) and (2) GNU Fortran (version 6.1 at the time of writing). The downloads will have the <code>.pkg</code> extension. Next, install <code>clang</code> and <code>gfortran</code> by opening these <code>.pkg</code> files and following the directions.</p>
<p>After this, add the following to your <code>~/.R/Makevars</code> file (replacing <code>clang8</code> with your version of the clang compiler):</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb10-1"><a href="#cb10-1"></a><span class="va">CLANG8=</span>/usr/local/clang8/bin/clang</span>
<span id="cb10-2"><a href="#cb10-2"></a><span class="va">CC=$(</span><span class="ex">CLANG8</span><span class="va">)</span></span>
<span id="cb10-3"><a href="#cb10-3"></a><span class="va">CXX=$(</span><span class="ex">CLANG8</span><span class="va">)</span>++</span>
<span id="cb10-4"><a href="#cb10-4"></a><span class="va">CXX11=$(</span><span class="ex">CLANG8</span><span class="va">)</span>++</span>
<span id="cb10-5"><a href="#cb10-5"></a><span class="va">CXX14=$(</span><span class="ex">CLANG8</span><span class="va">)</span>++</span>
<span id="cb10-6"><a href="#cb10-6"></a><span class="va">CXX17=$(</span><span class="ex">CLANG8</span><span class="va">)</span>++</span>
<span id="cb10-7"><a href="#cb10-7"></a><span class="va">CXX1X=$(</span><span class="ex">CLANG8</span><span class="va">)</span>++</span>
<span id="cb10-8"><a href="#cb10-8"></a><span class="va">LDFLAGS=</span>-L/usr/local/clang8/lib</span></code></pre></div>
<p>Now you should be able to install <code>jackalope</code> by running <code><a href="https://rdrr.io/r/utils/install.packages.html">install.packages("jackalope", type = "source")</a></code> in R.</p>
<p>For more information, please see <a href="https://thecoatlessprofessor.com/programming/openmp-in-r-on-os-x/" class="uri">https://thecoatlessprofessor.com/programming/openmp-in-r-on-os-x/</a>.</p>
</div>
</div>
</div>
<div id="usage" class="section level2">
<h2 class="hasAnchor">
<a href="#usage" class="anchor"></a>Usage</h2>
<p>Below shows how to simulate a 10kb genome, then create haplotypes from that genome using a phylogenetic tree:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/lucasnell/jackalope">jackalope</a></span><span class="op">)</span>
<span class="va">reference</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/create_genome.html">create_genome</a></span><span class="op">(</span>n_chroms <span class="op">=</span> <span class="fl">10</span>, len_mean <span class="op">=</span> <span class="fl">1000</span><span class="op">)</span>
<span class="va">tr</span> <span class="op">&lt;-</span> <span class="fu">ape</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/rtree.html">rcoal</a></span><span class="op">(</span><span class="fl">5</span><span class="op">)</span>
<span class="va">ref_haplotypes</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/create_haplotypes.html">create_haplotypes</a></span><span class="op">(</span><span class="va">reference</span>, <span class="fu"><a href="reference/haps_phylo.html">haps_phylo</a></span><span class="op">(</span><span class="va">tr</span><span class="op">)</span>, <span class="fu"><a href="reference/sub_models.html">sub_JC69</a></span><span class="op">(</span><span class="fl">0.1</span><span class="op">)</span><span class="op">)</span>
<span class="va">ref_haplotypes</span>
<span class="co">#&gt;                               &lt;&lt; haplotypes object &gt;&gt;</span>
<span class="co">#&gt; # Haplotypes: 5</span>
<span class="co">#&gt; # Mutations: 16,870</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;                           &lt;&lt; Reference genome info: &gt;&gt;</span>
<span class="co">#&gt; &lt; Set of 10 chromosomes &gt;</span>
<span class="co">#&gt; # Total size: 10,000 bp</span>
<span class="co">#&gt;   name                             chromosome                             length</span>
<span class="co">#&gt; chrom0     CTGGCATTGAATCATATGAGGTGGCCAT...ACGTTGCACGATTGATTAAATTCCTGAA      1000</span>
<span class="co">#&gt; chrom1     CACTCCGTCGCACACTAGGTTTCGAGAT...GTGAGCTCGCGTACATGGAGCATTCTGT      1000</span>
<span class="co">#&gt; chrom2     CTTAGCCGGAGCGACTCGGAGCAACTGC...TGGCGTAATATGCCAGGTCCCGCGTGGC      1000</span>
<span class="co">#&gt; chrom3     CGCCTTCCATTTAGGACTTGTATTGGTG...GCTAAACTCCATGTGACTGTAATGTCAG      1000</span>
<span class="co">#&gt; chrom4     GGGTGATATGGTGTGCATGCTGAATTCG...AGAGTCTAGAGTCTCTGGGAGGTCAGGT      1000</span>
<span class="co">#&gt; chrom5     TTCGTTGGTGGGTGTCCTATGCTACGAT...CGCCCGCCGGTTTGACTTACTCGATTGG      1000</span>
<span class="co">#&gt; chrom6     GCATGGACAGATGTGATCTGAGTATACG...CAGACCCCATAAGGCCTGGGACACTGTG      1000</span>
<span class="co">#&gt; chrom7     TCGTTTCAACGTCCTTAAGTGTAGTATC...GGCTCGTTAGCTCTCCGAGGAGACGAGG      1000</span>
<span class="co">#&gt; chrom8     CAGGTAAGTTATCAAAGAACCTTCCTGG...ACGCATCACCTCGCAAGGAGACTCGTTA      1000</span>
<span class="co">#&gt; chrom9     GGTAGTAATTAGGCTTAAAATAGCAGTG...ATAACAAATGTTCGGCATACGATCTACG      1000</span></code></pre></div>
<p>Below simulates 500 million paired-end, 100 bp reads from the haplotypes:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="reference/illumina.html">illumina</a></span><span class="op">(</span><span class="va">ref_haplotypes</span>, out_prefix <span class="op">=</span> <span class="st">"illumina"</span>, n_reads <span class="op">=</span> <span class="fl">500e6</span>,
         paired <span class="op">=</span> <span class="cn">TRUE</span>, read_length <span class="op">=</span> <span class="fl">100</span><span class="op">)</span></code></pre></div>
<p>Below simulates 500 thousand PacBio reads from the reference genome:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="reference/pacbio.html">pacbio</a></span><span class="op">(</span><span class="va">ref</span>, out_prefix <span class="op">=</span> <span class="st">"pacbio"</span>, n_reads <span class="op">=</span> <span class="fl">500e3</span><span class="op">)</span></code></pre></div>
</div>
</div>

  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2>Links</h2>
<ul class="list-unstyled">
<li>Download from CRAN at <br><a href="https://cloud.r-project.org/package=jackalope">https://​cloud.r-project.org/​package=jackalope</a>
</li>
<li>Browse source code at <br><a href="https://github.com/lucasnell/jackalope/">https://​github.com/​lucasnell/​jackalope/​</a>
</li>
<li>Report a bug at <br><a href="https://github.com/lucasnell/jackalope/issues">https://​github.com/​lucasnell/​jackalope/​issues</a>
</li>
</ul>
</div>
<div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li>
<a href="https://opensource.org/licenses/mit-license.php">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a>
</li>
</ul>
</div>
<div class="citation">
<h2>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html">Citing jackalope</a></li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>Lucas A. Nell <br><small class="roles"> Copyright holder, author, maintainer </small> <a href="https://orcid.org/0000-0003-3209-0517" target="orcid.widget" aria-label="ORCID"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
</ul>
</div>

  <div class="dev-status">
<h2>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://www.repostatus.org/#active"><img src="https://www.repostatus.org/badges/latest/active.svg" alt="Project Status: Active – The project has reached a stable, usable state and is being actively developed."></a></li>
<li><a href="https://github.com/lucasnell/jackalope/actions"><img src="https://github.com/lucasnell/jackalope/workflows/R-CMD-check/badge.svg" alt="R build status"></a></li>
<li><a href="https://codecov.io/gh/lucasnell/jackalope"><img src="https://codecov.io/gh/lucasnell/jackalope/branch/master/graph/badge.svg" alt="codecov"></a></li>
<li><a href="https://cran.r-project.org/package=jackalope"><img src="https://www.r-pkg.org/badges/version/jackalope" alt="CRAN_Status_Badge"></a></li>
</ul>
</div>
</div>
</div>


      <footer><div class="copyright">
  <p>Developed by Lucas A. Nell.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
