<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Create haplotypes from a reference genome. — create_haplotypes • jackalope</title><!-- favicons --><link rel="icon" type="image/png" sizes="48x48" href="../favicon-48x48.png"><link rel="icon" type="”image/svg+xml”" href="../favicon.svg"><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png"><link rel="icon" sizes="any" href="../favicon.ico"><link rel="manifest" href="../site.webmanifest"><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Create haplotypes from a reference genome. — create_haplotypes"><meta property="og:description" content="Uses one of multiple methods to create variant haplotypes from a reference genome.
See haps_functions for the methods available."><meta property="og:image" content="/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">jackalope</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.1.6</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/jackalope-intro.html">Intro to jackalope</a>
    </li>
    <li>
      <a href="../articles/sub-models.html">Models of nucleotide substitution</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/lucasnell/jackalope/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Create haplotypes from a reference genome.</h1>
    <small class="dont-index">Source: <a href="https://github.com/lucasnell/jackalope/blob/v1.1.6/R/create_haplotypes.R" class="external-link"><code>R/create_haplotypes.R</code></a></small>
    <div class="hidden name"><code>create_haplotypes.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Uses one of multiple methods to create variant haplotypes from a reference genome.
See <code><a href="haps_functions.html">haps_functions</a></code> for the methods available.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">create_haplotypes</span><span class="op">(</span></span>
<span>  <span class="va">reference</span>,</span>
<span>  <span class="va">haps_info</span>,</span>
<span>  sub <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  ins <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  del <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  epsilon <span class="op">=</span> <span class="fl">0.03</span>,</span>
<span>  n_threads <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  show_progress <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>


<dl><dt id="arg-reference">reference<a class="anchor" aria-label="anchor" href="#arg-reference"></a></dt>
<dd><p>A <code>ref_genome</code> object from which to generate haplotypes.
This argument is required.</p></dd>


<dt id="arg-haps-info">haps_info<a class="anchor" aria-label="anchor" href="#arg-haps-info"></a></dt>
<dd><p>Output from one of the <code><a href="haps_functions.html">haps_functions</a></code>.
These functions organize higher-level information for use here.
See <code><a href="haps_functions.html">haps_functions</a></code> for brief descriptions and links to each method.
If this argument is <code>NULL</code>, all arguments other than <code>reference</code> are ignored,
and an empty <code>haplotypes</code> object with no haplotypes is returned.
This is designed for use when you'd like to add mutations manually.
If you create a blank <code>haplotypes</code> object, you can use its <code>add_haps</code> method
to add haplotypes manually.</p></dd>


<dt id="arg-sub">sub<a class="anchor" aria-label="anchor" href="#arg-sub"></a></dt>
<dd><p>Output from one of the <code><a href="sub_models.html">sub_models</a></code> functions that organizes
information for the substitution models.
See <code><a href="sub_models.html">sub_models</a></code> for more information on these models and
their required parameters.
This argument is ignored if you are using a VCF file to create haplotypes.
Passing <code>NULL</code> to this argument results in no substitutions.
Defaults to <code>NULL</code>.</p></dd>


<dt id="arg-ins">ins<a class="anchor" aria-label="anchor" href="#arg-ins"></a></dt>
<dd><p>Output from the <code><a href="indels.html">indels</a></code> function that specifies rates
of insertions by length.
This argument is ignored if you are using a VCF file to create haplotypes.
Passing <code>NULL</code> to this argument results in no insertions.
Defaults to <code>NULL</code>.</p></dd>


<dt id="arg-del">del<a class="anchor" aria-label="anchor" href="#arg-del"></a></dt>
<dd><p>Output from the <code><a href="indels.html">indels</a></code> function that specifies rates
of deletions by length.
This argument is ignored if you are using a VCF file to create haplotypes.
Passing <code>NULL</code> to this argument results in no deletions.
Defaults to <code>NULL</code>.</p></dd>


<dt id="arg-epsilon">epsilon<a class="anchor" aria-label="anchor" href="#arg-epsilon"></a></dt>
<dd><p>Error control parameter for the "tau-leaping" approximation to
the Doob–Gillespie algorithm, as used for the indel portion of the simulations.
Smaller values result in a closer approximation.
Larger values are less exact but faster.
Values must be <code>&gt;= 0</code> and <code>&lt; 1</code>.
For more information on the approximation, see Cao et al. (2006) and
Wieder et al. (2011), listed below.
If <code>epsilon</code> is <code>0</code>, then it reverts to the exact Doob–Gillespie algorithm.
Defaults to <code>0.03</code>.</p></dd>


<dt id="arg-n-threads">n_threads<a class="anchor" aria-label="anchor" href="#arg-n-threads"></a></dt>
<dd><p>Number of threads to use for parallel processing.
This argument is ignored if OpenMP is not enabled.
Threads are spread across chromosomes, so it
doesn't make sense to supply more threads than chromosomes in the reference genome.
Defaults to <code>1</code>.</p></dd>


<dt id="arg-show-progress">show_progress<a class="anchor" aria-label="anchor" href="#arg-show-progress"></a></dt>
<dd><p>Boolean for whether to show a progress bar during processing.
Defaults to <code>FALSE</code>.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>A <code><a href="haplotypes.html">haplotypes</a></code> object.</p>
    </div>
    <div id="references">
    <h2>References</h2>
    <p>Cao, Y., D. T. Gillespie, and L. R. Petzold. 2006. Efficient step size
selection for the tau-leaping simulation method.
<em>The Journal of Chemical Physics</em> <strong>124</strong>(4): 044109.</p>
<p>Doob, J. L. 1942. Topics in the theory of markoff chains.
<em>Transactions of the American Mathematical Society</em> <strong>52</strong>(1): 37–64.</p>
<p>Gillespie, D. T. 1976. A general method for numerically simulating the stochastic time
evolution of coupled chemical reactions. <em>Journal of Computational Physics</em>
<strong>22</strong>(4): 403–434.</p>
<p>Wieder, N., R. H. Fink, and F. von Wegner. 2011. Exact and approximate stochastic
simulation of intracellular calcium dynamics.
<em>Journal of Biomedicine and Biotechnology</em> <strong>2011</strong>: 572492.</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">r</span> <span class="op">&lt;-</span> <span class="fu"><a href="create_genome.html">create_genome</a></span><span class="op">(</span><span class="fl">10</span>, <span class="fl">1000</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">v_phylo</span> <span class="op">&lt;-</span> <span class="fu">create_haplotypes</span><span class="op">(</span><span class="va">r</span>, <span class="fu"><a href="haps_phylo.html">haps_phylo</a></span><span class="op">(</span><span class="fu">ape</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/rtree.html" class="external-link">rcoal</a></span><span class="op">(</span><span class="fl">5</span><span class="op">)</span><span class="op">)</span>, <span class="fu"><a href="sub_models.html">sub_JC69</a></span><span class="op">(</span><span class="fl">0.1</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">v_theta</span> <span class="op">&lt;-</span> <span class="fu">create_haplotypes</span><span class="op">(</span><span class="va">r</span>, <span class="fu"><a href="haps_theta.html">haps_theta</a></span><span class="op">(</span><span class="fl">0.001</span>, <span class="fl">5</span><span class="op">)</span>, <span class="fu"><a href="sub_models.html">sub_K80</a></span><span class="op">(</span><span class="fl">0.1</span>, <span class="fl">0.2</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Lucas A. Nell.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

      </footer></div>






  </body></html>

