<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>An R6 Class Representing Haploid Variants — haplotypes • jackalope</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="An R6 Class Representing Haploid Variants — haplotypes"><meta property="og:description" content="Interactive wrapper for a pointer to a C++ object that stores information about
variant haplotypes from a single reference genome."><meta property="og:image" content="/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">jackalope</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.1.5</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/jackalope-intro.html">Intro to jackalope</a>
    </li>
    <li>
      <a href="../articles/sub-models.html">Models of nucleotide substitution</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/lucasnell/jackalope/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>An R6 Class Representing Haploid Variants</h1>
    <small class="dont-index">Source: <a href="https://github.com/lucasnell/jackalope/blob/HEAD/R/aaa-classes.R" class="external-link"><code>R/aaa-classes.R</code></a></small>
    <div class="hidden name"><code>haplotypes.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Interactive wrapper for a pointer to a C++ object that stores information about
variant haplotypes from a single reference genome.</p>
    </div>


    <div id="details">
    <h2>Details</h2>
    <p>This class should NEVER be created using <code>haplotypes$new</code>.
Only use <code>create_haplotypes</code>.
Because this class wraps a pointer to a C++ object, there are no fields to
manipulate directly.
All manipulations are done through this class's methods.</p>
    </div>
    <div id="connections-to-ref-genome-objects">
    <h2>Connections to <code>ref_genome</code> objects</h2>
    

<p>Regarding the <code>ref_genome</code> object you use to create a <code>haplotypes</code> object, you should
note the following:</p>
<ul><li><p><strong>This point is the most important.</strong>
Both the <code>ref_genome</code> and <code>haplotypes</code> objects use the same underlying
C++ object to store reference genome information.
Thus, if you make any changes to the <code>ref_genome</code> object, those changes will
also show up in the <code>haplotypes</code> object.
For example, if you make a <code>haplotypes</code> object named <code>V</code>
based on an existing <code>ref_genome</code> object named <code>R</code>,
then you merge chromosomes in <code>R</code>,
<code>V</code> will now have merged chromosomes.
If you've already started adding mutations to <code>V</code>,
then all the indexes used to store those mutations will be inaccurate.
So when you do anything with <code>V</code> later, your R session will crash
or have errors.
<strong>The lesson here is that you shouldn't edit the reference
genome after using it to create haplotypes.</strong></p></li>
<li><p>If a <code>ref_genome</code> object is used to create a <code>haplotypes</code>
object, deleting the <code>ref_genome</code> object won't cause issues with
the <code>haplotypes</code> object.
However, the <code>haplotypes</code> class doesn't provide methods to edit
chromosomes, so only remove the <code>ref_genome</code> object when you're done
editing the reference genome.</p></li>
</ul></div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p><code><a href="create_haplotypes.html">create_haplotypes</a></code></p></div>
    </div>
    <div id="methods">
    <h2>Methods</h2>
    
<div class="section">
<h3 id="public-methods">Public methods<a class="anchor" aria-label="anchor" href="#public-methods"></a></h3>

<ul><li><p><a href="#method-haplotypes-new"><code>haplotypes$new()</code></a></p></li>
<li><p><a href="#method-haplotypes-print"><code>haplotypes$print()</code></a></p></li>
<li><p><a href="#method-haplotypes-ptr"><code>haplotypes$ptr()</code></a></p></li>
<li><p><a href="#method-haplotypes-n_chroms"><code>haplotypes$n_chroms()</code></a></p></li>
<li><p><a href="#method-haplotypes-n_haps"><code>haplotypes$n_haps()</code></a></p></li>
<li><p><a href="#method-haplotypes-sizes"><code>haplotypes$sizes()</code></a></p></li>
<li><p><a href="#method-haplotypes-chrom_names"><code>haplotypes$chrom_names()</code></a></p></li>
<li><p><a href="#method-haplotypes-hap_names"><code>haplotypes$hap_names()</code></a></p></li>
<li><p><a href="#method-haplotypes-chrom"><code>haplotypes$chrom()</code></a></p></li>
<li><p><a href="#method-haplotypes-gc_prop"><code>haplotypes$gc_prop()</code></a></p></li>
<li><p><a href="#method-haplotypes-nt_prop"><code>haplotypes$nt_prop()</code></a></p></li>
<li><p><a href="#method-haplotypes-set_names"><code>haplotypes$set_names()</code></a></p></li>
<li><p><a href="#method-haplotypes-add_haps"><code>haplotypes$add_haps()</code></a></p></li>
<li><p><a href="#method-haplotypes-dup_haps"><code>haplotypes$dup_haps()</code></a></p></li>
<li><p><a href="#method-haplotypes-rm_haps"><code>haplotypes$rm_haps()</code></a></p></li>
<li><p><a href="#method-haplotypes-add_sub"><code>haplotypes$add_sub()</code></a></p></li>
<li><p><a href="#method-haplotypes-add_ins"><code>haplotypes$add_ins()</code></a></p></li>
<li><p><a href="#method-haplotypes-add_del"><code>haplotypes$add_del()</code></a></p></li>
</ul></div><p></p><hr><a id="method-haplotypes-new"></a><div class="section">
<h3 id="method-new-">Method <code>new()</code><a class="anchor" aria-label="anchor" href="#method-new-"></a></h3>
<p>Do NOT use this; only use <code>create_haplotypes</code> to make new <code>haplotypes</code>.</p><div class="section">
<h4 id="usage">Usage<a class="anchor" aria-label="anchor" href="#usage"></a></h4>
<p></p><div class="r"><div class="sourceCode"><pre><code><span><span class="va"><a href="../reference/haplotypes.html">haplotypes</a></span><span class="op">$</span><span class="fu">new</span><span class="op">(</span><span class="va">genomes_ptr</span>, <span class="va">reference_ptr</span><span class="op">)</span></span></code></pre></div><p></p></div>
</div>

<div class="section">
<h4 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h4>
<p></p><div class="arguments"><dl><dt><code>genomes_ptr</code></dt>
<dd><p>An <code>externalptr</code> object pointing to a C++ object that
stores the information about the haplotypes.</p></dd>


<dt><code>reference_ptr</code></dt>
<dd><p>An <code>externalptr</code> object pointing to a C++ object that
stores the information about the reference genome.</p></dd>


</dl><p></p></div>
</div>

</div><p></p><hr><a id="method-haplotypes-print"></a><div class="section">
<h3 id="method-print-">Method <code><a href="https://rdrr.io/r/base/print.html" class="external-link">print()</a></code><a class="anchor" aria-label="anchor" href="#method-print-"></a></h3>
<p>Print a <code>haplotypes</code> object.</p><div class="section">
<h4 id="usage-1">Usage<a class="anchor" aria-label="anchor" href="#usage-1"></a></h4>
<p></p><div class="r"><div class="sourceCode"><pre><code><span><span class="va">haplotypes</span><span class="op">$</span><span class="fu">print</span><span class="op">(</span><span class="op">)</span></span></code></pre></div><p></p></div>
</div>


</div><p></p><hr><a id="method-haplotypes-ptr"></a><div class="section">
<h3 id="method-ptr-">Method <code>ptr()</code><a class="anchor" aria-label="anchor" href="#method-ptr-"></a></h3>
<p>View pointer to underlying C++ object (this is not useful to end users).</p><div class="section">
<h4 id="usage-2">Usage<a class="anchor" aria-label="anchor" href="#usage-2"></a></h4>
<p></p><div class="r"><div class="sourceCode"><pre><code><span><span class="va">haplotypes</span><span class="op">$</span><span class="fu">ptr</span><span class="op">(</span><span class="op">)</span></span></code></pre></div><p></p></div>
</div>

<div class="section">
<h4 id="returns">Returns<a class="anchor" aria-label="anchor" href="#returns"></a></h4>
<p>An <code>externalptr</code> object.</p>
</div>

</div><p></p><hr><a id="method-haplotypes-n_chroms"></a><div class="section">
<h3 id="method-n-chroms-">Method <code>n_chroms()</code><a class="anchor" aria-label="anchor" href="#method-n-chroms-"></a></h3>
<p>View number of chromosomes.</p><div class="section">
<h4 id="usage-3">Usage<a class="anchor" aria-label="anchor" href="#usage-3"></a></h4>
<p></p><div class="r"><div class="sourceCode"><pre><code><span><span class="va">haplotypes</span><span class="op">$</span><span class="fu">n_chroms</span><span class="op">(</span><span class="op">)</span></span></code></pre></div><p></p></div>
</div>

<div class="section">
<h4 id="returns-1">Returns<a class="anchor" aria-label="anchor" href="#returns-1"></a></h4>
<p>Integer number of chromosomes.</p>
</div>

</div><p></p><hr><a id="method-haplotypes-n_haps"></a><div class="section">
<h3 id="method-n-haps-">Method <code>n_haps()</code><a class="anchor" aria-label="anchor" href="#method-n-haps-"></a></h3>
<p>View number of haplotypes.</p><div class="section">
<h4 id="usage-4">Usage<a class="anchor" aria-label="anchor" href="#usage-4"></a></h4>
<p></p><div class="r"><div class="sourceCode"><pre><code><span><span class="va">haplotypes</span><span class="op">$</span><span class="fu">n_haps</span><span class="op">(</span><span class="op">)</span></span></code></pre></div><p></p></div>
</div>

<div class="section">
<h4 id="returns-2">Returns<a class="anchor" aria-label="anchor" href="#returns-2"></a></h4>
<p>Integer number of haplotypes.</p>
</div>

</div><p></p><hr><a id="method-haplotypes-sizes"></a><div class="section">
<h3 id="method-sizes-">Method <code>sizes()</code><a class="anchor" aria-label="anchor" href="#method-sizes-"></a></h3>
<p>View chromosome sizes for one haplotype.</p><div class="section">
<h4 id="usage-5">Usage<a class="anchor" aria-label="anchor" href="#usage-5"></a></h4>
<p></p><div class="r"><div class="sourceCode"><pre><code><span><span class="va">haplotypes</span><span class="op">$</span><span class="fu">sizes</span><span class="op">(</span><span class="va">hap_ind</span><span class="op">)</span></span></code></pre></div><p></p></div>
</div>

<div class="section">
<h4 id="arguments-1">Arguments<a class="anchor" aria-label="anchor" href="#arguments-1"></a></h4>
<p></p><div class="arguments"><dl><dt><code>hap_ind</code></dt>
<dd><p>Index for the focal haplotype.</p></dd>


</dl><p></p></div>
</div>
<div class="section">
<h4 id="returns-3">Returns<a class="anchor" aria-label="anchor" href="#returns-3"></a></h4>
<p>Integer vector of chromosome sizes for focal haplotype.</p>
</div>

</div><p></p><hr><a id="method-haplotypes-chrom_names"></a><div class="section">
<h3 id="method-chrom-names-">Method <code>chrom_names()</code><a class="anchor" aria-label="anchor" href="#method-chrom-names-"></a></h3>
<p>View chromosome names.</p><div class="section">
<h4 id="usage-6">Usage<a class="anchor" aria-label="anchor" href="#usage-6"></a></h4>
<p></p><div class="r"><div class="sourceCode"><pre><code><span><span class="va">haplotypes</span><span class="op">$</span><span class="fu">chrom_names</span><span class="op">(</span><span class="op">)</span></span></code></pre></div><p></p></div>
</div>

<div class="section">
<h4 id="returns-4">Returns<a class="anchor" aria-label="anchor" href="#returns-4"></a></h4>
<p>Character vector of chromosome names.</p>
</div>

</div><p></p><hr><a id="method-haplotypes-hap_names"></a><div class="section">
<h3 id="method-hap-names-">Method <code>hap_names()</code><a class="anchor" aria-label="anchor" href="#method-hap-names-"></a></h3>
<p>View haplotype names.</p><div class="section">
<h4 id="usage-7">Usage<a class="anchor" aria-label="anchor" href="#usage-7"></a></h4>
<p></p><div class="r"><div class="sourceCode"><pre><code><span><span class="va">haplotypes</span><span class="op">$</span><span class="fu">hap_names</span><span class="op">(</span><span class="op">)</span></span></code></pre></div><p></p></div>
</div>

<div class="section">
<h4 id="returns-5">Returns<a class="anchor" aria-label="anchor" href="#returns-5"></a></h4>
<p>Character vector of haplotype names.</p>
</div>

</div><p></p><hr><a id="method-haplotypes-chrom"></a><div class="section">
<h3 id="method-chrom-">Method <code>chrom()</code><a class="anchor" aria-label="anchor" href="#method-chrom-"></a></h3>
<p>View one haplotype chromosome.</p><div class="section">
<h4 id="usage-8">Usage<a class="anchor" aria-label="anchor" href="#usage-8"></a></h4>
<p></p><div class="r"><div class="sourceCode"><pre><code><span><span class="va">haplotypes</span><span class="op">$</span><span class="fu">chrom</span><span class="op">(</span><span class="va">hap_ind</span>, <span class="va">chrom_ind</span><span class="op">)</span></span></code></pre></div><p></p></div>
</div>

<div class="section">
<h4 id="arguments-2">Arguments<a class="anchor" aria-label="anchor" href="#arguments-2"></a></h4>
<p></p><div class="arguments"><dl><dt><code>hap_ind</code></dt>
<dd><p>Index for the focal haplotype.</p></dd>


<dt><code>chrom_ind</code></dt>
<dd><p>Index for the focal chromosome.</p></dd>


</dl><p></p></div>
</div>
<div class="section">
<h4 id="returns-6">Returns<a class="anchor" aria-label="anchor" href="#returns-6"></a></h4>
<p>A single string representing the chosen haplotype chromosome's DNA
sequence.</p>
</div>

</div><p></p><hr><a id="method-haplotypes-gc_prop"></a><div class="section">
<h3 id="method-gc-prop-">Method <code>gc_prop()</code><a class="anchor" aria-label="anchor" href="#method-gc-prop-"></a></h3>
<p>View GC proportion for part of one haplotype chromosome.</p><div class="section">
<h4 id="usage-9">Usage<a class="anchor" aria-label="anchor" href="#usage-9"></a></h4>
<p></p><div class="r"><div class="sourceCode"><pre><code><span><span class="va">haplotypes</span><span class="op">$</span><span class="fu">gc_prop</span><span class="op">(</span><span class="va">hap_ind</span>, <span class="va">chrom_ind</span>, <span class="va">start</span>, <span class="va">end</span><span class="op">)</span></span></code></pre></div><p></p></div>
</div>

<div class="section">
<h4 id="arguments-3">Arguments<a class="anchor" aria-label="anchor" href="#arguments-3"></a></h4>
<p></p><div class="arguments"><dl><dt><code>hap_ind</code></dt>
<dd><p>Index for the focal haplotype.</p></dd>


<dt><code>chrom_ind</code></dt>
<dd><p>Index for the focal chromosome.</p></dd>


<dt><code>start</code></dt>
<dd><p>Point on the chromosome at which to start the calculation
(inclusive).</p></dd>


<dt><code>end</code></dt>
<dd><p>Point on the chromosome at which to end the calculation
(inclusive).</p></dd>


</dl><p></p></div>
</div>
<div class="section">
<h4 id="returns-7">Returns<a class="anchor" aria-label="anchor" href="#returns-7"></a></h4>
<p>A double in the range <code>[0,1]</code> representing the proportion of DNA
sequence that is either <code>G</code> or <code>C</code>.</p>
</div>

</div><p></p><hr><a id="method-haplotypes-nt_prop"></a><div class="section">
<h3 id="method-nt-prop-">Method <code>nt_prop()</code><a class="anchor" aria-label="anchor" href="#method-nt-prop-"></a></h3>
<p>View nucleotide content for part of one haplotype chromosome</p><div class="section">
<h4 id="usage-10">Usage<a class="anchor" aria-label="anchor" href="#usage-10"></a></h4>
<p></p><div class="r"><div class="sourceCode"><pre><code><span><span class="va">haplotypes</span><span class="op">$</span><span class="fu">nt_prop</span><span class="op">(</span><span class="va">nt</span>, <span class="va">hap_ind</span>, <span class="va">chrom_ind</span>, <span class="va">start</span>, <span class="va">end</span><span class="op">)</span></span></code></pre></div><p></p></div>
</div>

<div class="section">
<h4 id="arguments-4">Arguments<a class="anchor" aria-label="anchor" href="#arguments-4"></a></h4>
<p></p><div class="arguments"><dl><dt><code>nt</code></dt>
<dd><p>Which nucleotide to calculate the proportion that the DNA
sequence is made of. Must be one of <code>T</code>, <code>C</code>, <code>A</code>, <code>G</code>, or <code>N</code>.</p></dd>


<dt><code>hap_ind</code></dt>
<dd><p>Index for the focal haplotype.</p></dd>


<dt><code>chrom_ind</code></dt>
<dd><p>Index for the focal chromosome.</p></dd>


<dt><code>start</code></dt>
<dd><p>Point on the chromosome at which to start the calculation
(inclusive).</p></dd>


<dt><code>end</code></dt>
<dd><p>Point on the chromosome at which to end the calculation
(inclusive).</p></dd>


</dl><p></p></div>
</div>
<div class="section">
<h4 id="returns-8">Returns<a class="anchor" aria-label="anchor" href="#returns-8"></a></h4>
<p>A double in the range <code>[0,1]</code> representing the proportion of DNA
sequence that is <code>nt</code>.</p>
</div>

</div><p></p><hr><a id="method-haplotypes-set_names"></a><div class="section">
<h3 id="method-set-names-">Method <code>set_names()</code><a class="anchor" aria-label="anchor" href="#method-set-names-"></a></h3>
<p>Change haplotype names.</p><div class="section">
<h4 id="usage-11">Usage<a class="anchor" aria-label="anchor" href="#usage-11"></a></h4>
<p></p><div class="r"><div class="sourceCode"><pre><code><span><span class="va">haplotypes</span><span class="op">$</span><span class="fu">set_names</span><span class="op">(</span><span class="va">new_names</span><span class="op">)</span></span></code></pre></div><p></p></div>
</div>

<div class="section">
<h4 id="arguments-5">Arguments<a class="anchor" aria-label="anchor" href="#arguments-5"></a></h4>
<p></p><div class="arguments"><dl><dt><code>new_names</code></dt>
<dd><p>Vector of new names to use. This must be the same length as
the number of current names.</p></dd>


</dl><p></p></div>
</div>
<div class="section">
<h4 id="returns-9">Returns<a class="anchor" aria-label="anchor" href="#returns-9"></a></h4>
<p>This <code>R6</code> object, invisibly.</p>
</div>

</div><p></p><hr><a id="method-haplotypes-add_haps"></a><div class="section">
<h3 id="method-add-haps-">Method <code>add_haps()</code><a class="anchor" aria-label="anchor" href="#method-add-haps-"></a></h3>
<p>Add one or more blank, named haplotypes</p><div class="section">
<h4 id="usage-12">Usage<a class="anchor" aria-label="anchor" href="#usage-12"></a></h4>
<p></p><div class="r"><div class="sourceCode"><pre><code><span><span class="va">haplotypes</span><span class="op">$</span><span class="fu">add_haps</span><span class="op">(</span><span class="va">new_names</span><span class="op">)</span></span></code></pre></div><p></p></div>
</div>

<div class="section">
<h4 id="arguments-6">Arguments<a class="anchor" aria-label="anchor" href="#arguments-6"></a></h4>
<p></p><div class="arguments"><dl><dt><code>new_names</code></dt>
<dd><p>Vector of name(s) for the new haplotype(s).</p></dd>


</dl><p></p></div>
</div>
<div class="section">
<h4 id="returns-10">Returns<a class="anchor" aria-label="anchor" href="#returns-10"></a></h4>
<p>This <code>R6</code> object, invisibly.</p>
</div>

</div><p></p><hr><a id="method-haplotypes-dup_haps"></a><div class="section">
<h3 id="method-dup-haps-">Method <code>dup_haps()</code><a class="anchor" aria-label="anchor" href="#method-dup-haps-"></a></h3>
<p>Duplicate one or more haplotypes by name.</p><div class="section">
<h4 id="usage-13">Usage<a class="anchor" aria-label="anchor" href="#usage-13"></a></h4>
<p></p><div class="r"><div class="sourceCode"><pre><code><span><span class="va">haplotypes</span><span class="op">$</span><span class="fu">dup_haps</span><span class="op">(</span><span class="va">hap_names</span>, new_names <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span></span></code></pre></div><p></p></div>
</div>

<div class="section">
<h4 id="arguments-7">Arguments<a class="anchor" aria-label="anchor" href="#arguments-7"></a></h4>
<p></p><div class="arguments"><dl><dt><code>hap_names</code></dt>
<dd><p>Vector of existing haplotype name(s) that you want to
duplicate.</p></dd>


<dt><code>new_names</code></dt>
<dd><p>Optional vector specifying the names of the duplicates.
If <code>NULL</code>, their names are auto-generated. Defaults to <code>NULL</code>.</p></dd>


</dl><p></p></div>
</div>
<div class="section">
<h4 id="returns-11">Returns<a class="anchor" aria-label="anchor" href="#returns-11"></a></h4>
<p>This <code>R6</code> object, invisibly.</p>
</div>

</div><p></p><hr><a id="method-haplotypes-rm_haps"></a><div class="section">
<h3 id="method-rm-haps-">Method <code>rm_haps()</code><a class="anchor" aria-label="anchor" href="#method-rm-haps-"></a></h3>
<p>Remove one or more haplotypes by name.</p><div class="section">
<h4 id="usage-14">Usage<a class="anchor" aria-label="anchor" href="#usage-14"></a></h4>
<p></p><div class="r"><div class="sourceCode"><pre><code><span><span class="va">haplotypes</span><span class="op">$</span><span class="fu">rm_haps</span><span class="op">(</span><span class="va">hap_names</span><span class="op">)</span></span></code></pre></div><p></p></div>
</div>

<div class="section">
<h4 id="arguments-8">Arguments<a class="anchor" aria-label="anchor" href="#arguments-8"></a></h4>
<p></p><div class="arguments"><dl><dt><code>hap_names</code></dt>
<dd><p>Vector of existing haplotype name(s) that you want to remove.</p></dd>


</dl><p></p></div>
</div>
<div class="section">
<h4 id="returns-12">Returns<a class="anchor" aria-label="anchor" href="#returns-12"></a></h4>
<p>This <code>R6</code> object, invisibly.</p>
</div>

</div><p></p><hr><a id="method-haplotypes-add_sub"></a><div class="section">
<h3 id="method-add-sub-">Method <code>add_sub()</code><a class="anchor" aria-label="anchor" href="#method-add-sub-"></a></h3>
<p>Manually add a substitution.</p><div class="section">
<h4 id="usage-15">Usage<a class="anchor" aria-label="anchor" href="#usage-15"></a></h4>
<p></p><div class="r"><div class="sourceCode"><pre><code><span><span class="va">haplotypes</span><span class="op">$</span><span class="fu">add_sub</span><span class="op">(</span><span class="va">hap_ind</span>, <span class="va">chrom_ind</span>, <span class="va">pos</span>, <span class="va">nt</span><span class="op">)</span></span></code></pre></div><p></p></div>
</div>

<div class="section">
<h4 id="arguments-9">Arguments<a class="anchor" aria-label="anchor" href="#arguments-9"></a></h4>
<p></p><div class="arguments"><dl><dt><code>hap_ind</code></dt>
<dd><p>Index for the focal haplotype.</p></dd>


<dt><code>chrom_ind</code></dt>
<dd><p>Index for the focal chromosome.</p></dd>


<dt><code>pos</code></dt>
<dd><p>Position at which to add the mutation.</p></dd>


<dt><code>nt</code></dt>
<dd><p>Single character representing the nucleotide to change the
current one to.</p></dd>


</dl><p></p></div>
</div>
<div class="section">
<h4 id="returns-13">Returns<a class="anchor" aria-label="anchor" href="#returns-13"></a></h4>
<p>This <code>R6</code> object, invisibly.</p>
</div>

</div><p></p><hr><a id="method-haplotypes-add_ins"></a><div class="section">
<h3 id="method-add-ins-">Method <code>add_ins()</code><a class="anchor" aria-label="anchor" href="#method-add-ins-"></a></h3>
<p>Manually add an insertion.</p><div class="section">
<h4 id="usage-16">Usage<a class="anchor" aria-label="anchor" href="#usage-16"></a></h4>
<p></p><div class="r"><div class="sourceCode"><pre><code><span><span class="va">haplotypes</span><span class="op">$</span><span class="fu">add_ins</span><span class="op">(</span><span class="va">hap_ind</span>, <span class="va">chrom_ind</span>, <span class="va">pos</span>, <span class="va">nts</span><span class="op">)</span></span></code></pre></div><p></p></div>
</div>

<div class="section">
<h4 id="arguments-10">Arguments<a class="anchor" aria-label="anchor" href="#arguments-10"></a></h4>
<p></p><div class="arguments"><dl><dt><code>hap_ind</code></dt>
<dd><p>Index for the focal haplotype.</p></dd>


<dt><code>chrom_ind</code></dt>
<dd><p>Index for the focal chromosome.</p></dd>


<dt><code>pos</code></dt>
<dd><p>Position at which to add the mutation.</p></dd>


<dt><code>nts</code></dt>
<dd><p>String representing the nucleotide(s) that will be inserted after
the designated position.</p></dd>


</dl><p></p></div>
</div>
<div class="section">
<h4 id="returns-14">Returns<a class="anchor" aria-label="anchor" href="#returns-14"></a></h4>
<p>This <code>R6</code> object, invisibly.</p>
<p></p><div class="sourceCode"><pre><code><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a>\item{<span class="st">`</span><span class="at">add_del(hap_ind, chrom_ind, pos, n_nts)</span><span class="st">`</span>}{Manually add a deletion</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> a given <span class="fu">haplotype</span> (<span class="st">`</span><span class="at">hap_ind</span><span class="st">`</span>), <span class="fu">chromosome</span> (<span class="st">`</span><span class="at">chrom_ind</span><span class="st">`</span>), and <span class="fu">position</span> (<span class="st">`</span><span class="at">pos</span><span class="st">`</span>).</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a>    The designated number of nucleotides to <span class="fu">delete</span> (<span class="st">`</span><span class="at">n_nts</span><span class="st">`</span>) will be deleted</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a>    starting at <span class="st">`</span><span class="at">pos</span><span class="st">`</span>, unless <span class="st">`</span><span class="at">pos</span><span class="st">`</span> is near the chromosome end and doesn<span class="st">'t have</span></span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="st">    `n_nts` nucleotides to remove; it simply stops at the chromosome end in</span></span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="st">    this case.}</span></span></code></pre><p></p></div>
</div>

</div><p></p><hr><a id="method-haplotypes-add_del"></a><div class="section">
<h3 id="method-add-del-">Method <code>add_del()</code><a class="anchor" aria-label="anchor" href="#method-add-del-"></a></h3>
<p>Manually add a deletion.</p><div class="section">
<h4 id="usage-17">Usage<a class="anchor" aria-label="anchor" href="#usage-17"></a></h4>
<p></p><div class="r"><div class="sourceCode"><pre><code><span><span class="va">haplotypes</span><span class="op">$</span><span class="fu">add_del</span><span class="op">(</span><span class="va">hap_ind</span>, <span class="va">chrom_ind</span>, <span class="va">pos</span>, <span class="va">n_nts</span><span class="op">)</span></span></code></pre></div><p></p></div>
</div>

<div class="section">
<h4 id="arguments-11">Arguments<a class="anchor" aria-label="anchor" href="#arguments-11"></a></h4>
<p></p><div class="arguments"><dl><dt><code>hap_ind</code></dt>
<dd><p>Index for the focal haplotype.</p></dd>


<dt><code>chrom_ind</code></dt>
<dd><p>Index for the focal chromosome.</p></dd>


<dt><code>pos</code></dt>
<dd><p>Position at which to add the mutation.</p></dd>


<dt><code>n_nts</code></dt>
<dd><p>Single integer specifying the number of nucleotides to delete.
These will be deleted starting at <code>pos</code>.
If <code>pos</code> is near the chromosome end and doesn't have <code>n_nts</code> nucleotides
to remove, it simply removes nucleotides from <code>pos</code> to the chromosome end.</p></dd>


</dl><p></p></div>
</div>
<div class="section">
<h4 id="returns-15">Returns<a class="anchor" aria-label="anchor" href="#returns-15"></a></h4>
<p>This <code>R6</code> object, invisibly.</p>
</div>

</div>

    </div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Lucas A. Nell.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

