<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Intro to jackalope • jackalope</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Intro to jackalope">
<meta property="og:description" content="">
<meta property="og:image" content="/logo.png">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">jackalope</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.0.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/jackalope-intro.html">Intro to jackalope</a>
    </li>
    <li>
      <a href="../articles/sub-models.html">Models of nucleotide substitution</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/lucasnell/jackalope">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Intro to jackalope</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/lucasnell/jackalope/blob/master/vignettes/jackalope-intro.Rmd"><code>vignettes/jackalope-intro.Rmd</code></a></small>
      <div class="hidden name"><code>jackalope-intro.Rmd</code></div>

    </div>

    
    
<p>This document provides brief examples of how <code>jackalope</code> can be used to generate sequencing data that can inform some common sampling decisions for HTS studies.</p>
<p>For an example reference assembly, I simply simulated one of size 1 kb using the following code:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1">ref &lt;-<span class="st"> </span><span class="kw"><a href="../reference/create_genome.html">create_genome</a></span>(<span class="dv">10</span>, <span class="fl">1e2</span>)</a></code></pre></div>
<p>This resulted in the following <code>ref_genome</code> object:</p>
<pre><code>#&gt; &lt; Set of 10 chromosomes &gt;
#&gt; # Total size: 1,000 bp
#&gt;   name                             chromosome                             length
#&gt; chrom0     TGTTCTCACGGAAAACGGGAAGTTTTTG...GTCACGTCAATGGAGTCTTGGGTCTCTA       100
#&gt; chrom1     CCTCATGTTTCTAGTAGCTGCAACGCTC...TTGTGAACTCCCAGAGGCCTATGCTAAA       100
#&gt; chrom2     AAATCGTCAAGGCTTCGACGTAGAGCCG...CATTGCAGTTCGGAACCGAGGACGGGCG       100
#&gt; chrom3     TTCAGGCAGTGTGGTCTACTGGGCGCAG...TCAGAGCGATAAGATAGATCCGGATAAT       100
#&gt; chrom4     AGGCTGACACCGCCGATGGCCGGGTCCA...GGGCGTCGGGCGAATGCTCGCATGCGAA       100
#&gt; chrom5     CACTGAACGTTAGCAACTTCATCTACAT...GATTTCCTATCTGTGTGTTAGGACTATA       100
#&gt; chrom6     GCGGTTAAGACCGAACCATCCGTGGGAC...AAAGTCTCTCGCAAGCCAGTCAGTGCCC       100
#&gt; chrom7     CCGAATGCCAGTAAGCTCGAGCGTATAA...GTAAGCGTAACGCGCGTGCTGGTTTCGT       100
#&gt; chrom8     TATAGCTAGACCGCCTCCGCGTACGACC...CGGCCGGGCTATCGGGCCTGGTAGGGGG       100
#&gt; chrom9     TAAAGAGTGGGCGGGTCCGTAAAGTACC...CCACGAAGGCTCACGGTCACAATAGGTT       100</code></pre>
<p>For molecular-evolution information, I used the TN93 model, an insertion rate of <code>2e-5</code> for lengths from 1 to 10, and a deletion rate of <code>1e-5</code> for lengths from 1 to 40. By default, the <code>indel</code> function creates rates that are proportional to <code><a href="https://rdrr.io/r/base/Log.html">exp(-L)</a></code> for indel length <code>L</code> from 1 to the maximum length.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" data-line-number="1">sub &lt;-<span class="st"> </span><span class="kw"><a href="../reference/sub_models.html">sub_TN93</a></span>(<span class="dt">pi_tcag =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">0.1</span>, <span class="fl">0.2</span>, <span class="fl">0.3</span>, <span class="fl">0.4</span>),</a>
<a class="sourceLine" id="cb3-2" data-line-number="2">                <span class="dt">alpha_1 =</span> <span class="fl">0.0001</span>, <span class="dt">alpha_2 =</span> <span class="fl">0.0002</span>,</a>
<a class="sourceLine" id="cb3-3" data-line-number="3">                <span class="dt">beta =</span> <span class="fl">0.00015</span>)</a>
<a class="sourceLine" id="cb3-4" data-line-number="4">ins &lt;-<span class="st"> </span><span class="kw"><a href="../reference/indels.html">indels</a></span>(<span class="dt">rate =</span> <span class="fl">2e-5</span>, <span class="dt">max_length =</span> <span class="dv">10</span>)</a>
<a class="sourceLine" id="cb3-5" data-line-number="5">del &lt;-<span class="st"> </span><span class="kw"><a href="../reference/indels.html">indels</a></span>(<span class="dt">rate =</span> <span class="fl">1e-5</span>, <span class="dt">max_length =</span> <span class="dv">40</span>)</a></code></pre></div>
<div id="assembling-a-genome" class="section level2">
<h2 class="hasAnchor">
<a href="#assembling-a-genome" class="anchor"></a>Assembling a genome</h2>
<p>The example here produces FASTQ files from the known reference assembly that could test strategies for how to assemble a similar genome using HTS data.</p>
<p>The first strategy is to use only PacBio sequencing. The PacBio Sequel system produces up to 500,000 reads per Single Molecule, Real-Time (SMRT) cell, so you could run the following for two cells:</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" data-line-number="1"><span class="kw"><a href="../reference/pacbio.html">pacbio</a></span>(ref, <span class="dt">out_prefix =</span> <span class="st">"pacbio"</span>, <span class="dt">n_reads =</span> <span class="dv">2</span> <span class="op">*</span><span class="st"> </span><span class="fl">500e3</span>)</a></code></pre></div>
<p>An alternative, hybrid strategy uses 1 SMRT cell of PacBio sequencing and 1 lane (<span class="math inline">\(\sim 500\)</span> million reads) of <span class="math inline">\(2 \times 100\)</span>bp Illumina sequencing on the HiSeq 2500 system (the default Illumina system in <code>jackalope</code>):</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" data-line-number="1"><span class="kw"><a href="../reference/pacbio.html">pacbio</a></span>(ref, <span class="dt">out_prefix =</span> <span class="st">"pacbio"</span>, <span class="dt">n_reads =</span> <span class="fl">500e3</span>)</a>
<a class="sourceLine" id="cb5-2" data-line-number="2"><span class="kw"><a href="../reference/illumina.html">illumina</a></span>(ref, <span class="dt">out_prefix =</span> <span class="st">"illumina"</span>, <span class="dt">n_reads =</span> <span class="fl">500e6</span>, <span class="dt">paired =</span> <span class="ot">TRUE</span>,</a>
<a class="sourceLine" id="cb5-3" data-line-number="3">         <span class="dt">read_length =</span> <span class="dv">100</span>)</a></code></pre></div>
<p>The last strategy combines 1 lane of <span class="math inline">\(2 \times 100\)</span>bp Illumina HiSeq 2500 sequencing with 1 flow cell of <span class="math inline">\(2 \times 250\)</span>bp mate-pair sequencing on an Illumina MiSeq v3. The mate-pair sequencing uses longer fragments (defaults are mean of 400 and standard deviation of 100) to better cover highly repetitive regions.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" data-line-number="1"><span class="kw"><a href="../reference/illumina.html">illumina</a></span>(ref, <span class="dt">out_prefix =</span> <span class="st">"ill_pe"</span>, <span class="dt">n_reads =</span> <span class="fl">500e6</span>, <span class="dt">paired =</span> <span class="ot">TRUE</span>,</a>
<a class="sourceLine" id="cb6-2" data-line-number="2">         <span class="dt">read_length =</span> <span class="dv">100</span>)</a>
<a class="sourceLine" id="cb6-3" data-line-number="3"><span class="kw"><a href="../reference/illumina.html">illumina</a></span>(ref, <span class="dt">out_prefix =</span> <span class="st">"ill_mp"</span>, <span class="dt">seq_sys =</span> <span class="st">"MSv3"</span>,</a>
<a class="sourceLine" id="cb6-4" data-line-number="4">         <span class="dt">read_length =</span> <span class="dv">250</span>, <span class="dt">n_reads =</span> <span class="fl">50e6</span>, <span class="dt">matepair =</span> <span class="ot">TRUE</span>, </a>
<a class="sourceLine" id="cb6-5" data-line-number="5">         <span class="dt">frag_mean =</span> <span class="dv">3000</span>, <span class="dt">frag_sd =</span> <span class="dv">500</span>)</a></code></pre></div>
</div>
<div id="estimating-divergence-between-populations" class="section level2">
<h2 class="hasAnchor">
<a href="#estimating-divergence-between-populations" class="anchor"></a>Estimating divergence between populations</h2>
<p>Here, I will demonstrate how to generate population-genomic data of a type that might be used to estimate the divergence between two populations. I first use the <code>scrm</code> package to conduct coalescent simulations that will generate segregating sites for 10 variant haplotypes from the reference genome. Five of the haplotypes are from one population, five from another. The symmetrical migration rate is 100 individuals per generation. I set the population-scaled mutation rate (<span class="math inline">\(\theta = 4 N_0 \mu\)</span>) to <code>10</code> for this example.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb7-1" data-line-number="1"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(scrm)</a>
<a class="sourceLine" id="cb7-2" data-line-number="2">ssites &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/scrm/man/scrm.html">scrm</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/paste.html">paste</a></span>(<span class="st">"10"</span>, ref<span class="op">$</span><span class="kw">n_chroms</span>(), <span class="st">"-t 10 -I 2 5 5 100"</span>))</a></code></pre></div>
<p>Using the previously created objects for molecular evolution information (<code>sub</code>, <code>ins</code>, and <code>del</code>), I create haplotypes from the reference genome:</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb8-1" data-line-number="1">haps &lt;-<span class="st"> </span><span class="kw"><a href="../reference/create_haplotypes.html">create_haplotypes</a></span>(ref, <span class="kw"><a href="../reference/haps_ssites.html">haps_ssites</a></span>(ssites), sub, ins, del)</a></code></pre></div>
<p>This results in the following set of haplotypes:</p>
<pre><code>#&gt;                               &lt;&lt; haplotypes object &gt;&gt;
#&gt; # Haplotypes: 10
#&gt; # Mutations: 471
#&gt; 
#&gt;                           &lt;&lt; Reference genome info: &gt;&gt;
#&gt; &lt; Set of 10 chromosomes &gt;
#&gt; # Total size: 1,000 bp
#&gt;   name                             chromosome                             length
#&gt; chrom0     TGTTCTCACGGAAAACGGGAAGTTTTTG...GTCACGTCAATGGAGTCTTGGGTCTCTA       100
#&gt; chrom1     CCTCATGTTTCTAGTAGCTGCAACGCTC...TTGTGAACTCCCAGAGGCCTATGCTAAA       100
#&gt; chrom2     AAATCGTCAAGGCTTCGACGTAGAGCCG...CATTGCAGTTCGGAACCGAGGACGGGCG       100
#&gt; chrom3     TTCAGGCAGTGTGGTCTACTGGGCGCAG...TCAGAGCGATAAGATAGATCCGGATAAT       100
#&gt; chrom4     AGGCTGACACCGCCGATGGCCGGGTCCA...GGGCGTCGGGCGAATGCTCGCATGCGAA       100
#&gt; chrom5     CACTGAACGTTAGCAACTTCATCTACAT...GATTTCCTATCTGTGTGTTAGGACTATA       100
#&gt; chrom6     GCGGTTAAGACCGAACCATCCGTGGGAC...AAAGTCTCTCGCAAGCCAGTCAGTGCCC       100
#&gt; chrom7     CCGAATGCCAGTAAGCTCGAGCGTATAA...GTAAGCGTAACGCGCGTGCTGGTTTCGT       100
#&gt; chrom8     TATAGCTAGACCGCCTCCGCGTACGACC...CGGCCGGGCTATCGGGCCTGGTAGGGGG       100
#&gt; chrom9     TAAAGAGTGGGCGGGTCCGTAAAGTACC...CCACGAAGGCTCACGGTCACAATAGGTT       100</code></pre>
<p>For a file of true divergences from the reference genome, the <code>write_vcf</code> function writes the <code>haplotypes</code> object to a VCF file:</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb10-1" data-line-number="1"><span class="kw"><a href="../reference/write_vcf.html">write_vcf</a></span>(haps, <span class="st">"haplotypes"</span>)</a></code></pre></div>
<p>Lastly, I simulate 1 lane of <span class="math inline">\(2 \times 100\)</span>bp Illumina HiSeq 2500 sequencing. In this case, individuals within a population are pooled, and the population sequences are derived from are identified by barcodes.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb11-1" data-line-number="1"><span class="kw"><a href="../reference/illumina.html">illumina</a></span>(haps, <span class="dt">out_prefix =</span> <span class="st">"haps_illumina"</span>, <span class="dt">n_reads =</span> <span class="fl">500e6</span>, <span class="dt">paired =</span> <span class="ot">TRUE</span>,</a>
<a class="sourceLine" id="cb11-2" data-line-number="2">         <span class="dt">read_length =</span> <span class="dv">100</span>, <span class="dt">barcodes =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/rep.html">rep</a></span>(<span class="st">"AACCGCGG"</span>, <span class="dv">20</span>), </a>
<a class="sourceLine" id="cb11-3" data-line-number="3">                                         <span class="kw"><a href="https://rdrr.io/r/base/rep.html">rep</a></span>(<span class="st">"GGTTATAA"</span>, <span class="dv">20</span>)))</a></code></pre></div>
<p>The below example instead has each individual haplotype’s reads output to separate FASTQ files:</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb12-1" data-line-number="1"><span class="kw"><a href="../reference/illumina.html">illumina</a></span>(haps, <span class="dt">out_prefix =</span> <span class="st">"haps_illumina"</span>, <span class="dt">n_reads =</span> <span class="fl">500e6</span>, <span class="dt">paired =</span> <span class="ot">TRUE</span>,</a>
<a class="sourceLine" id="cb12-2" data-line-number="2">         <span class="dt">read_length =</span> <span class="dv">100</span>, <span class="dt">sep_files =</span> <span class="ot">TRUE</span>)</a></code></pre></div>
</div>
<div id="constructing-a-phylogeny" class="section level2">
<h2 class="hasAnchor">
<a href="#constructing-a-phylogeny" class="anchor"></a>Constructing a phylogeny</h2>
<div id="from-one-phylogenetic-tree" class="section level3">
<h3 class="hasAnchor">
<a href="#from-one-phylogenetic-tree" class="anchor"></a>From one phylogenetic tree</h3>
<p>This section shows how <code>jackalope</code> can generate haplotypes from a phylogeny, then simulate sequencing data from those haplotypes to test phylogeny reconstruction methods. First, I generated a random coalescent tree of 10 species using the <code>ape</code> package.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb13-1" data-line-number="1">tree &lt;-<span class="st"> </span><span class="kw">rcoal</span>(<span class="dv">10</span>)</a></code></pre></div>
<p>Then I input that to the <code>create_haplotypes</code> function alongside the molecular evolution information.</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb14-1" data-line-number="1">haps &lt;-<span class="st"> </span><span class="kw"><a href="../reference/create_haplotypes.html">create_haplotypes</a></span>(ref, <span class="kw"><a href="../reference/haps_phylo.html">haps_phylo</a></span>(tree), sub, ins, del)</a></code></pre></div>
<p>This results in the following <code>haplotypes</code> object:</p>
<pre><code>#&gt;                               &lt;&lt; haplotypes object &gt;&gt;
#&gt; # Haplotypes: 10
#&gt; # Mutations: 3
#&gt; 
#&gt;                           &lt;&lt; Reference genome info: &gt;&gt;
#&gt; &lt; Set of 10 chromosomes &gt;
#&gt; # Total size: 1,000 bp
#&gt;   name                             chromosome                             length
#&gt; chrom0     TGTTCTCACGGAAAACGGGAAGTTTTTG...GTCACGTCAATGGAGTCTTGGGTCTCTA       100
#&gt; chrom1     CCTCATGTTTCTAGTAGCTGCAACGCTC...TTGTGAACTCCCAGAGGCCTATGCTAAA       100
#&gt; chrom2     AAATCGTCAAGGCTTCGACGTAGAGCCG...CATTGCAGTTCGGAACCGAGGACGGGCG       100
#&gt; chrom3     TTCAGGCAGTGTGGTCTACTGGGCGCAG...TCAGAGCGATAAGATAGATCCGGATAAT       100
#&gt; chrom4     AGGCTGACACCGCCGATGGCCGGGTCCA...GGGCGTCGGGCGAATGCTCGCATGCGAA       100
#&gt; chrom5     CACTGAACGTTAGCAACTTCATCTACAT...GATTTCCTATCTGTGTGTTAGGACTATA       100
#&gt; chrom6     GCGGTTAAGACCGAACCATCCGTGGGAC...AAAGTCTCTCGCAAGCCAGTCAGTGCCC       100
#&gt; chrom7     CCGAATGCCAGTAAGCTCGAGCGTATAA...GTAAGCGTAACGCGCGTGCTGGTTTCGT       100
#&gt; chrom8     TATAGCTAGACCGCCTCCGCGTACGACC...CGGCCGGGCTATCGGGCCTGGTAGGGGG       100
#&gt; chrom9     TAAAGAGTGGGCGGGTCCGTAAAGTACC...CCACGAAGGCTCACGGTCACAATAGGTT       100</code></pre>
<p>Now I can generate data for 1 flow cell of <span class="math inline">\(2 \times 250\)</span>bp sequencing on an Illumina MiSeq v3.</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb16-1" data-line-number="1"><span class="kw"><a href="../reference/illumina.html">illumina</a></span>(haps, <span class="dt">out_prefix =</span> <span class="st">"phylo_tree"</span>, <span class="dt">seq_sys =</span> <span class="st">"MSv3"</span>,</a>
<a class="sourceLine" id="cb16-2" data-line-number="2">         <span class="dt">read_length =</span> <span class="dv">250</span>, <span class="dt">n_reads =</span> <span class="fl">50e6</span>)</a></code></pre></div>
</div>
<div id="from-gene-trees" class="section level3">
<h3 class="hasAnchor">
<a href="#from-gene-trees" class="anchor"></a>From gene trees</h3>
<p>Similar to the section above, the ultimate goal here is to test phylogeny reconstruction methods. The difference in this section is that instead of using a single, straightforward phylogeny, I use multiple gene trees per chromosome. In the species used in these simulations, species 3 diverged from 1 and 2 at <span class="math inline">\(t = 0.5\)</span>, where <span class="math inline">\(t\)</span> indicates time into the past and is in units of <span class="math inline">\(4 N_0\)</span> generations. Species 1 and 2 diverged at <span class="math inline">\(t = 0.25\)</span>. I assume a recombination rate of <span class="math inline">\(1 / (4 N_0)\)</span> recombination events per chromosome per generation. Because each chromosome is a different length and the length is required for including a recombination rate, I had to run <code>scrm</code> separately for each chromosome. There are 4 diploid individuals sampled per species.</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb17-1" data-line-number="1"><span class="co"># Run scrm for one chromosome size:</span></a>
<a class="sourceLine" id="cb17-2" data-line-number="2">one_chrom &lt;-<span class="st"> </span><span class="cf">function</span>(size) {</a>
<a class="sourceLine" id="cb17-3" data-line-number="3">    sims &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/scrm/man/scrm.html">scrm</a></span>(</a>
<a class="sourceLine" id="cb17-4" data-line-number="4">        <span class="kw"><a href="https://rdrr.io/r/base/paste.html">paste</a></span>(<span class="st">"24 1"</span>,</a>
<a class="sourceLine" id="cb17-5" data-line-number="5">              <span class="co"># Output gene trees:</span></a>
<a class="sourceLine" id="cb17-6" data-line-number="6">              <span class="st">"-T"</span>,</a>
<a class="sourceLine" id="cb17-7" data-line-number="7">              <span class="co"># Recombination:</span></a>
<a class="sourceLine" id="cb17-8" data-line-number="8">              <span class="st">"-r 1"</span>, size,</a>
<a class="sourceLine" id="cb17-9" data-line-number="9">              <span class="co"># 3 species with no ongoing migration:</span></a>
<a class="sourceLine" id="cb17-10" data-line-number="10">              <span class="st">"-I 3"</span>, <span class="kw"><a href="https://rdrr.io/r/base/paste.html">paste</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/rep.html">rep</a></span>(<span class="st">"8"</span>, <span class="dv">3</span>), <span class="dt">collapse =</span> <span class="st">" "</span>), <span class="st">"0"</span>,</a>
<a class="sourceLine" id="cb17-11" data-line-number="11">              <span class="co"># Species 2 derived from 1 at time 1.0:</span></a>
<a class="sourceLine" id="cb17-12" data-line-number="12">              <span class="st">"-ej 0.5 2 1"</span>,  </a>
<a class="sourceLine" id="cb17-13" data-line-number="13">              <span class="co"># Species 3 derived from 2 at time 0.5:</span></a>
<a class="sourceLine" id="cb17-14" data-line-number="14">              <span class="st">"-ej 0.25 3 2"</span></a>
<a class="sourceLine" id="cb17-15" data-line-number="15">        ))</a>
<a class="sourceLine" id="cb17-16" data-line-number="16">    <span class="kw"><a href="https://rdrr.io/r/base/function.html">return</a></span>(sims<span class="op">$</span>trees[[<span class="dv">1</span>]])</a>
<a class="sourceLine" id="cb17-17" data-line-number="17">}</a>
<a class="sourceLine" id="cb17-18" data-line-number="18"><span class="co"># For all chromosomes:</span></a>
<a class="sourceLine" id="cb17-19" data-line-number="19">gtrees &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="dt">trees =</span> <span class="kw"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span>(ref<span class="op">$</span><span class="kw">sizes</span>(), one_chrom))</a></code></pre></div>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb18-1" data-line-number="1">haps &lt;-<span class="st"> </span><span class="kw"><a href="../reference/create_haplotypes.html">create_haplotypes</a></span>(ref, <span class="kw"><a href="../reference/haps_gtrees.html">haps_gtrees</a></span>(gtrees),</a>
<a class="sourceLine" id="cb18-2" data-line-number="2">                        sub, ins, del)</a></code></pre></div>
<p>This results in the following <code>haplotypes</code> object:</p>
<pre><code>#&gt;                               &lt;&lt; haplotypes object &gt;&gt;
#&gt; # Haplotypes: 24
#&gt; # Mutations: 117
#&gt; 
#&gt;                           &lt;&lt; Reference genome info: &gt;&gt;
#&gt; &lt; Set of 10 chromosomes &gt;
#&gt; # Total size: 1,000 bp
#&gt;   name                             chromosome                             length
#&gt; chrom0     TGTTCTCACGGAAAACGGGAAGTTTTTG...GTCACGTCAATGGAGTCTTGGGTCTCTA       100
#&gt; chrom1     CCTCATGTTTCTAGTAGCTGCAACGCTC...TTGTGAACTCCCAGAGGCCTATGCTAAA       100
#&gt; chrom2     AAATCGTCAAGGCTTCGACGTAGAGCCG...CATTGCAGTTCGGAACCGAGGACGGGCG       100
#&gt; chrom3     TTCAGGCAGTGTGGTCTACTGGGCGCAG...TCAGAGCGATAAGATAGATCCGGATAAT       100
#&gt; chrom4     AGGCTGACACCGCCGATGGCCGGGTCCA...GGGCGTCGGGCGAATGCTCGCATGCGAA       100
#&gt; chrom5     CACTGAACGTTAGCAACTTCATCTACAT...GATTTCCTATCTGTGTGTTAGGACTATA       100
#&gt; chrom6     GCGGTTAAGACCGAACCATCCGTGGGAC...AAAGTCTCTCGCAAGCCAGTCAGTGCCC       100
#&gt; chrom7     CCGAATGCCAGTAAGCTCGAGCGTATAA...GTAAGCGTAACGCGCGTGCTGGTTTCGT       100
#&gt; chrom8     TATAGCTAGACCGCCTCCGCGTACGACC...CGGCCGGGCTATCGGGCCTGGTAGGGGG       100
#&gt; chrom9     TAAAGAGTGGGCGGGTCCGTAAAGTACC...CCACGAAGGCTCACGGTCACAATAGGTT       100</code></pre>
<p>To store mutation information by diploid sample, the <code>write_vcf</code> function writes the <code>haplotypes</code> object to a VCF file. It assigns haplotypes to diploid samples using a matrix for the <code>sample_matrix</code> argument:</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb20-1" data-line-number="1"><span class="kw"><a href="../reference/write_vcf.html">write_vcf</a></span>(haps, <span class="dt">out_prefix =</span> <span class="st">"hap_gtrees"</span>,</a>
<a class="sourceLine" id="cb20-2" data-line-number="2">          <span class="dt">sample_matrix =</span> <span class="kw"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span>(<span class="dv">1</span><span class="op">:</span>haps<span class="op">$</span><span class="kw">n_haps</span>(), <span class="dt">ncol =</span> <span class="dv">2</span>, <span class="dt">byrow =</span> <span class="ot">TRUE</span>))</a></code></pre></div>
<p>Now I can generate data for 1 flow cell of <span class="math inline">\(2 \times 250\)</span>bp sequencing on an Illumina MiSeq v3.</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb21-1" data-line-number="1"><span class="kw"><a href="../reference/illumina.html">illumina</a></span>(ref, <span class="dt">out_prefix =</span> <span class="st">"phylo_gtrees"</span>,</a>
<a class="sourceLine" id="cb21-2" data-line-number="2">         <span class="dt">seq_sys =</span> <span class="st">"MSv3"</span>,</a>
<a class="sourceLine" id="cb21-3" data-line-number="3">         <span class="dt">paired =</span> <span class="ot">TRUE</span>,</a>
<a class="sourceLine" id="cb21-4" data-line-number="4">         <span class="dt">read_length =</span> <span class="dv">250</span>,</a>
<a class="sourceLine" id="cb21-5" data-line-number="5">         <span class="dt">n_reads =</span> <span class="fl">50e6</span>)</a></code></pre></div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">

        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#assembling-a-genome">Assembling a genome</a></li>
      <li><a href="#estimating-divergence-between-populations">Estimating divergence between populations</a></li>
      <li><a href="#constructing-a-phylogeny">Constructing a phylogeny</a></li>
      </ul>
</div>
      </div>

</div>



      <footer><div class="copyright">
  <p>Developed by Lucas A. Nell.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
