<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Intro to jackalope • jackalope</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Intro to jackalope">
<meta property="og:description" content="jackalope">
<meta property="og:image" content="/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">jackalope</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.1.4</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/jackalope-intro.html">Intro to jackalope</a>
    </li>
    <li>
      <a href="../articles/sub-models.html">Models of nucleotide substitution</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/lucasnell/jackalope/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Intro to jackalope</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/lucasnell/jackalope/blob/HEAD/vignettes/jackalope-intro.Rmd" class="external-link"><code>vignettes/jackalope-intro.Rmd</code></a></small>
      <div class="hidden name"><code>jackalope-intro.Rmd</code></div>

    </div>

    
    
<p>This document provides brief examples of how <code>jackalope</code>
can be used to generate sequencing data that can inform some common
sampling decisions for HTS studies.</p>
<div class="section level2">
<h2 id="generating-reference-genome">Generating reference genome<a class="anchor" aria-label="anchor" href="#generating-reference-genome"></a>
</h2>
<p>Here, I’ll show how to generate a reference genome based on an
existing assembly or via simulated DNA sequences.</p>
<p>To show how an existing assembly is read in <code>jackalope</code>,
the code below processes the assembly for <em>Drosophila
melanogaster</em> (version 6.27) downloaded from <a href="https://flybase.org" class="external-link uri">https://flybase.org</a>. It reads
the compressed FASTA file, filters out scaffolds by using a size
threshold, removes the Y chromosome (to avoid having both an X and Y in
the same haplotype), and merges the left and right arms of chromosomes 2
and 3. It lastly sets the names of chromosomes 2 and 3 to be
<code>"2"</code> and <code>"3"</code>, respectively.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ref</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/read_fasta.html">read_fasta</a></span><span class="op">(</span><span class="st">"dmel-6.27.fasta.gz"</span>, cut_names <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">ref</span><span class="op">$</span><span class="fu">filter_chroms</span><span class="op">(</span><span class="fl">1e6</span>, method <span class="op">=</span> <span class="st">"size"</span><span class="op">)</span></span>
<span><span class="va">ref</span><span class="op">$</span><span class="fu">rm_chroms</span><span class="op">(</span><span class="st">"Y"</span><span class="op">)</span></span>
<span><span class="va">ref</span><span class="op">$</span><span class="fu">merge_chroms</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"2L"</span>, <span class="st">"2R"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">ref</span><span class="op">$</span><span class="fu">merge_chroms</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"3L"</span>, <span class="st">"3R"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">names</span> <span class="op">&lt;-</span> <span class="va">ref</span><span class="op">$</span><span class="fu">chrom_names</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">names</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grepl</a></span><span class="op">(</span><span class="st">"^2"</span>, <span class="va">names</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"2"</span></span>
<span><span class="va">names</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grepl</a></span><span class="op">(</span><span class="st">"^3"</span>, <span class="va">names</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"3"</span></span>
<span><span class="va">ref</span><span class="op">$</span><span class="fu">set_names</span><span class="op">(</span><span class="va">names</span><span class="op">)</span></span></code></pre></div>
<p>For the rest of the document, I will use a simulated genome for
simplicity (and so that I don’t have to store the <em>D.
melanogaster</em> genome inside this package). Here is how I simulated a
genome of size <span class="math inline">\(\sim 1\)</span> kb split
among 4 chromosomes, with the same names as the chromosomes in the
<em>D. melanogaster</em> genome:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ref</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/create_genome.html">create_genome</a></span><span class="op">(</span>n_chroms <span class="op">=</span> <span class="fl">4</span>,</span>
<span>                     len_mean <span class="op">=</span> <span class="fl">1e3</span> <span class="op">/</span> <span class="fl">4</span>,</span>
<span>                     len_sd <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span>
<span><span class="va">ref</span><span class="op">$</span><span class="fu">set_names</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span><span class="op">:</span><span class="fl">4</span>, <span class="st">"X"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>This resulted in the following <code>ref_genome</code> object:</p>
<pre><code><span><span class="co">#&gt; &lt; Set of 4 chromosomes &gt;</span></span>
<span><span class="co">#&gt; # Total size: 985 bp</span></span>
<span><span class="co">#&gt;   name                             chromosome                             length</span></span>
<span><span class="co">#&gt; 2          GAAATCATGTCTTATAAAACGCCAAATC...TCCCCGCTTTCAAGACAGTCTCCTCGAC       252</span></span>
<span><span class="co">#&gt; 3          TCCGTGATCATGCACAAGCATTGCAGTT...TCTAACGGGCGTCGGGCGAATGCTCGCA       240</span></span>
<span><span class="co">#&gt; 4          CACGCGAGCTTACTATTGGTAATGCCTT...CACTTCGGAAGTCAGTTACCAGCTCCTC       253</span></span>
<span><span class="co">#&gt; X          AGTCGATCGCAGACCGTCCTCCTGTAAG...ACACGTCACAGCCACGAAGGCTCACGGT       240</span></span></code></pre>
<p>For the examples below, we’ll pretend this is our <em>D.
melanogaster</em> genome.</p>
</div>
<div class="section level2">
<h2 id="molecular-evolution-information">Molecular evolution information<a class="anchor" aria-label="anchor" href="#molecular-evolution-information"></a>
</h2>
<p>To generate variant haplotypes based on our reference genome, we need
molecular evolution information (unless passing a Variant Call Format,
or VCF, file). For molecular-evolution information, I used the JC69
model for simplicity. Mutation rates were chosen from the
<code>evo_rates</code> object present inside <code>jackalope</code>,
which stores Table 1 from <a href="https://dx.doi.org/10.1534/g3.116.030890" class="external-link">Sung et al. (2016)</a>.
The substitution and indel rates were the values from
<code>evo_rates</code> specific to <em>D. melanogaster</em>. The
<code>mu</code> parameter to <code>sub_JC69</code> function was set to
<code>NULL</code> because the default behavior of all substitution-model
functions in <code>jackalope</code> is to scale rate matrices such that
branch lengths are in units of substitutions per site. In this case, I
don’t want to scale rate matrices because our branch lengths will be in
generations. <a href="https://doi.org/10.1093/nar/gkg745" class="external-link">Zhang and
Gerstein (2003)</a> found a 1 to 2.90 ratio of insertions to deletions,
and described relative rates of indels of various sizes using a
power-law relationship. To approximate the distributions they described,
relative rates were derived from a Lavalette distribution with <span class="math inline">\(L = 60\)</span> and <span class="math inline">\(a
= 1.60\)</span> for insertions and <span class="math inline">\(a =
1.51\)</span> for deletions. The <span class="math inline">\(\theta\)</span> parameter is the population-scaled
mutation rate, which we can get directly from <code>evo_rates</code> for
<em>D. melanogaster</em>. Lastly, <span class="math inline">\(N_0\)</span> is the effective population size for
<em>D. melanogaster</em>, which is used in a few instances below because
<code>scrm</code> outputs branch lengths in units of <span class="math inline">\(4 N_0\)</span> generations.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sub_rate</span> <span class="op">&lt;-</span> <span class="va">evo_rates</span><span class="op">$</span><span class="va">subs</span><span class="op">[</span><span class="va">evo_rates</span><span class="op">$</span><span class="va">species</span> <span class="op">==</span> <span class="st">"Drosophila melanogaster"</span><span class="op">]</span></span>
<span><span class="va">indel_rate</span> <span class="op">&lt;-</span> <span class="va">evo_rates</span><span class="op">$</span><span class="va">indels</span><span class="op">[</span><span class="va">evo_rates</span><span class="op">$</span><span class="va">species</span> <span class="op">==</span> <span class="st">"Drosophila melanogaster"</span><span class="op">]</span></span>
<span><span class="co"># Because both are in units of 10^-10 events per site per generation:</span></span>
<span><span class="va">sub_rate</span> <span class="op">&lt;-</span> <span class="va">sub_rate</span> <span class="op">*</span> <span class="fl">1e-10</span></span>
<span><span class="va">indel_rate</span> <span class="op">&lt;-</span> <span class="va">indel_rate</span> <span class="op">*</span> <span class="fl">1e-10</span></span>
<span></span>
<span><span class="va">sub</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sub_models.html">sub_JC69</a></span><span class="op">(</span>lambda <span class="op">=</span> <span class="va">sub_rate</span>, mu <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span></span>
<span><span class="va">ins</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/indels.html">indels</a></span><span class="op">(</span>rate <span class="op">=</span> <span class="va">indel_rate</span>, max_length <span class="op">=</span> <span class="fl">60</span>,a <span class="op">=</span> <span class="fl">1.60</span><span class="op">)</span></span>
<span><span class="va">del</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/indels.html">indels</a></span><span class="op">(</span>rate <span class="op">=</span> <span class="va">indel_rate</span>, max_length <span class="op">=</span> <span class="fl">60</span>, a <span class="op">=</span> <span class="fl">1.51</span><span class="op">)</span></span>
<span></span>
<span><span class="va">theta</span> <span class="op">&lt;-</span> <span class="va">evo_rates</span><span class="op">[</span><span class="va">evo_rates</span><span class="op">$</span><span class="va">species</span> <span class="op">==</span> <span class="st">"Drosophila melanogaster"</span>,<span class="st">"theta_s"</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># Originally in units of 1e6 individuals</span></span>
<span><span class="va">N0</span> <span class="op">&lt;-</span> <span class="va">evo_rates</span><span class="op">[</span><span class="va">evo_rates</span><span class="op">$</span><span class="va">species</span> <span class="op">==</span> <span class="st">"Drosophila melanogaster"</span>, <span class="st">"Ne"</span><span class="op">]</span> <span class="op">*</span> <span class="fl">1e6</span></span></code></pre></div>
<p>When generating any haplotypes below, these objects will be used
inside <code>create_haplotypes</code> to specify molecular evolution
information.</p>
</div>
<div class="section level2">
<h2 id="assembling-a-genome">Assembling a genome<a class="anchor" aria-label="anchor" href="#assembling-a-genome"></a>
</h2>
<div class="section level3">
<h3 id="based-on-a-reference">Based on a reference<a class="anchor" aria-label="anchor" href="#based-on-a-reference"></a>
</h3>
<p>The examples here produce FASTQ files from the known reference
assembly that could test strategies for how to assemble a similar genome
using HTS data.</p>
<p>The first strategy is to use only PacBio sequencing. The PacBio
Sequel system produces up to 500,000 reads per Single Molecule,
Real-Time (SMRT) cell, so you could run the following for two cells
(with the file <code>pacbio_R1.fq</code> as output):</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/pacbio.html">pacbio</a></span><span class="op">(</span><span class="va">ref</span>, out_prefix <span class="op">=</span> <span class="st">"pacbio"</span>, n_reads <span class="op">=</span> <span class="fl">2</span> <span class="op">*</span> <span class="fl">500e3</span><span class="op">)</span></span></code></pre></div>
<p>An alternative, hybrid strategy uses 1 SMRT cell of PacBio sequencing
and 1 lane (<span class="math inline">\(\sim 500\)</span> million reads)
of <span class="math inline">\(2 \times 100\)</span>bp Illumina
sequencing on the HiSeq 2500 system (the default Illumina system in
<code>jackalope</code>):</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/pacbio.html">pacbio</a></span><span class="op">(</span><span class="va">ref</span>, out_prefix <span class="op">=</span> <span class="st">"pacbio"</span>, n_reads <span class="op">=</span> <span class="fl">500e3</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/illumina.html">illumina</a></span><span class="op">(</span><span class="va">ref</span>, out_prefix <span class="op">=</span> <span class="st">"illumina"</span>, n_reads <span class="op">=</span> <span class="fl">500e6</span>, paired <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>         read_length <span class="op">=</span> <span class="fl">100</span><span class="op">)</span></span></code></pre></div>
<p>The last strategy combines 1 lane of <span class="math inline">\(2
\times 100\)</span>bp Illumina HiSeq 2500 sequencing with 1 flow cell of
<span class="math inline">\(2 \times 250\)</span>bp mate-pair sequencing
on an Illumina MiSeq v3. The mate-pair sequencing uses longer fragments
(defaults are mean of 400 and standard deviation of 100) to better cover
highly repetitive regions.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/illumina.html">illumina</a></span><span class="op">(</span><span class="va">ref</span>, out_prefix <span class="op">=</span> <span class="st">"ill_pe"</span>, n_reads <span class="op">=</span> <span class="fl">500e6</span>, paired <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>         read_length <span class="op">=</span> <span class="fl">100</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/illumina.html">illumina</a></span><span class="op">(</span><span class="va">ref</span>, out_prefix <span class="op">=</span> <span class="st">"ill_mp"</span>, seq_sys <span class="op">=</span> <span class="st">"MSv3"</span>,</span>
<span>         read_length <span class="op">=</span> <span class="fl">250</span>, n_reads <span class="op">=</span> <span class="fl">50e6</span>, matepair <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>         frag_mean <span class="op">=</span> <span class="fl">3000</span>, frag_sd <span class="op">=</span> <span class="fl">500</span><span class="op">)</span></span></code></pre></div>
<p>These data could then be used to compare genome assembly performance
between the strategies above, or between programs within a given
strategy. Extensions of these tests include adjusting sequencing depth,
sequencing platform, and error rates.</p>
</div>
<div class="section level3">
<h3 id="based-on-a-diploid-individual">Based on a diploid individual<a class="anchor" aria-label="anchor" href="#based-on-a-diploid-individual"></a>
</h3>
<p>For diploid species, scientists won’t be sampling a haploid
reference, and heterozygosity can be a real problem for the assembly
process. So below I’ll show you how to simulate a diploid individual and
create reads based on that. I’ll just show the hybrid assembly strategy,
but the others simply differ in the sequencing step as shown above.</p>
<p>First, we’ll simulate two haplotypes based on the <span class="math inline">\(\theta\)</span> (population-scaled mutation rate)
and the other molecular evolution information for <em>D.
melanogaster</em> we got from the <code>evo_rates</code> object earlier.
I’m also renaming the haplotypes.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">haps</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/create_haplotypes.html">create_haplotypes</a></span><span class="op">(</span><span class="va">ref</span>, <span class="fu"><a href="../reference/haps_theta.html">haps_theta</a></span><span class="op">(</span>theta <span class="op">=</span> <span class="va">theta</span>, n_haps <span class="op">=</span> <span class="fl">2</span><span class="op">)</span>, </span>
<span>                          <span class="va">sub</span>, <span class="va">ins</span>, <span class="va">del</span><span class="op">)</span></span>
<span><span class="va">haps</span><span class="op">$</span><span class="fu">set_names</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"A"</span>, <span class="st">"B"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>This results in the following <code>haplotypes</code> object:</p>
<pre><code><span><span class="co">#&gt;                               &lt;&lt; haplotypes object &gt;&gt;</span></span>
<span><span class="co">#&gt; # Haplotypes: 2</span></span>
<span><span class="co">#&gt; # Mutations: 15</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;                           &lt;&lt; Reference genome info: &gt;&gt;</span></span>
<span><span class="co">#&gt; &lt; Set of 4 chromosomes &gt;</span></span>
<span><span class="co">#&gt; # Total size: 985 bp</span></span>
<span><span class="co">#&gt;   name                             chromosome                             length</span></span>
<span><span class="co">#&gt; 2          GAAATCATGTCTTATAAAACGCCAAATC...TCCCCGCTTTCAAGACAGTCTCCTCGAC       252</span></span>
<span><span class="co">#&gt; 3          TCCGTGATCATGCACAAGCATTGCAGTT...TCTAACGGGCGTCGGGCGAATGCTCGCA       240</span></span>
<span><span class="co">#&gt; 4          CACGCGAGCTTACTATTGGTAATGCCTT...CACTTCGGAAGTCAGTTACCAGCTCCTC       253</span></span>
<span><span class="co">#&gt; X          AGTCGATCGCAGACCGTCCTCCTGTAAG...ACACGTCACAGCCACGAAGGCTCACGGT       240</span></span></code></pre>
<p>We generate 1 SMRT cell of PacBio sequencing and 1 lane of <span class="math inline">\(2 \times 100\)</span>bp Illumina sequencing on the
HiSeq 2500 system as before, except using the new <code>haps</code>
object instead of <code>ref</code>:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/pacbio.html">pacbio</a></span><span class="op">(</span><span class="va">haps</span>, out_prefix <span class="op">=</span> <span class="st">"pacbio"</span>, n_reads <span class="op">=</span> <span class="fl">500e3</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/illumina.html">illumina</a></span><span class="op">(</span><span class="va">haps</span>, out_prefix <span class="op">=</span> <span class="st">"illumina"</span>, n_reads <span class="op">=</span> <span class="fl">500e6</span>, paired <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>         read_length <span class="op">=</span> <span class="fl">100</span><span class="op">)</span></span></code></pre></div>
<p>If we want to save the information for these haplotypes, we can
output FASTA files (one per haplotype) or a VCF file:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/write_fasta.html">write_fasta</a></span><span class="op">(</span><span class="va">haps</span>, <span class="st">"haps"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/write_vcf.html">write_vcf</a></span><span class="op">(</span><span class="va">haps</span>, <span class="st">"haps"</span>, sample_matrix <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>This would result in the following files being generated:
<code>haps__A.fa</code>, <code>haps__B.fa</code>, and
<code>haps.vcf</code>. The <code>sample_matrix</code> argument to
<code>write_vcf</code> states that the first two (i.e., all) haplotypes
are from the first (and only) sample. These data could test different
assembly strategies, similar to the above section, except that, for
diploid organisms, it includes heterozygosity. Increasing the <span class="math inline">\(\theta\)</span> parameter will result in more
heterozygosity, which is an obvious extension of these tests.</p>
</div>
</div>
<div class="section level2">
<h2 id="estimating-divergence-between-populations">Estimating divergence between populations<a class="anchor" aria-label="anchor" href="#estimating-divergence-between-populations"></a>
</h2>
<p>Here, I will demonstrate how to generate population-genomic data of a
type that might be used to estimate the divergence between two
populations. I first use the <code>scrm</code> package to conduct
coalescent simulations that will generate segregating sites for 10
variant haplotypes from the reference genome. Five of the haplotypes are
from one population, five from another. The symmetrical migration rate
is 100 individuals per generation. I used a recombination rate of 1 (the
expected number of recombinations on the locus per <span class="math inline">\(4 N_0\)</span> generations). I specify the
recombination rate and chromosome size for <code>scrm</code> using the
command <code>-r 1 C</code> for chromosome size <code>C</code>. I used
the population-scaled mutation rate for <em>D. melanogaster</em>, scaled
such that the mutation rate (<span class="math inline">\(\mu\)</span>)
is in units of mutations per <span class="math inline">\(4 N_0\)</span>
generations, where <span class="math inline">\(N_0\)</span> is the
population size. For <span class="math inline">\(N_0\)</span>, I’m using
the effective population size for <em>D. melanogaster</em>, per
<code>evo_rates</code>. The other code in the chunk below is to convert
the output to look like that from the <code>scrm</code> R package, which
can be easily processed by <code>jackalope</code>.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/scrm/scrm-r" class="external-link">scrm</a></span><span class="op">)</span></span>
<span><span class="co"># Function to run scrm for one chromosome and format output</span></span>
<span><span class="va">one_chrom</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">.size</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">ssites</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scrm/man/scrm.html" class="external-link">scrm</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sprintf.html" class="external-link">sprintf</a></span><span class="op">(</span><span class="st">"10 1 -t %.4f -r 1 %i -I 2 5 5 100"</span>, <span class="va">theta</span> <span class="op">*</span> <span class="fl">4</span> <span class="op">*</span> <span class="va">N0</span>, <span class="va">.size</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="va">ssites</span><span class="op">$</span><span class="va">seg_sites</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="va">ssites</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>seg_sites <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">ref</span><span class="op">$</span><span class="fu">sizes</span><span class="op">(</span><span class="op">)</span>, <span class="va">one_chrom</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Using the previously created objects for molecular evolution
information and the <code>haps_ssites</code> function, I create
haplotypes from the reference genome:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">haps</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/create_haplotypes.html">create_haplotypes</a></span><span class="op">(</span><span class="va">ref</span>, <span class="fu"><a href="../reference/haps_ssites.html">haps_ssites</a></span><span class="op">(</span><span class="va">ssites</span><span class="op">)</span>, <span class="va">sub</span>, <span class="va">ins</span>, <span class="va">del</span><span class="op">)</span></span></code></pre></div>
<p>This results in the following set of haplotypes:</p>
<pre><code><span><span class="co">#&gt;                               &lt;&lt; haplotypes object &gt;&gt;</span></span>
<span><span class="co">#&gt; # Haplotypes: 10</span></span>
<span><span class="co">#&gt; # Mutations: 2,682</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;                           &lt;&lt; Reference genome info: &gt;&gt;</span></span>
<span><span class="co">#&gt; &lt; Set of 4 chromosomes &gt;</span></span>
<span><span class="co">#&gt; # Total size: 985 bp</span></span>
<span><span class="co">#&gt;   name                             chromosome                             length</span></span>
<span><span class="co">#&gt; 2          GAAATCATGTCTTATAAAACGCCAAATC...TCCCCGCTTTCAAGACAGTCTCCTCGAC       252</span></span>
<span><span class="co">#&gt; 3          TCCGTGATCATGCACAAGCATTGCAGTT...TCTAACGGGCGTCGGGCGAATGCTCGCA       240</span></span>
<span><span class="co">#&gt; 4          CACGCGAGCTTACTATTGGTAATGCCTT...CACTTCGGAAGTCAGTTACCAGCTCCTC       253</span></span>
<span><span class="co">#&gt; X          AGTCGATCGCAGACCGTCCTCCTGTAAG...ACACGTCACAGCCACGAAGGCTCACGGT       240</span></span></code></pre>
<p>For a file of true divergences from the reference genome, the
<code>write_vcf</code> function writes the <code>haplotypes</code>
object to a VCF file:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/write_vcf.html">write_vcf</a></span><span class="op">(</span><span class="va">haps</span>, <span class="st">"haplotypes"</span><span class="op">)</span></span></code></pre></div>
<p>Lastly, I simulate 1 lane of <span class="math inline">\(2 \times
100\)</span>bp Illumina HiSeq 2500 sequencing. In this case, individuals
within a population are pooled, and the population sequences are derived
from are identified by barcodes.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/illumina.html">illumina</a></span><span class="op">(</span><span class="va">haps</span>, out_prefix <span class="op">=</span> <span class="st">"haps_illumina"</span>, n_reads <span class="op">=</span> <span class="fl">500e6</span>, paired <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>         read_length <span class="op">=</span> <span class="fl">100</span>, barcodes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"AACCGCGG"</span>, <span class="fl">5</span><span class="op">)</span>, </span>
<span>                                         <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"GGTTATAA"</span>, <span class="fl">5</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>The below example instead has each individual haplotype’s reads
output to separate FASTQ files:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/illumina.html">illumina</a></span><span class="op">(</span><span class="va">haps</span>, out_prefix <span class="op">=</span> <span class="st">"haps_illumina"</span>, n_reads <span class="op">=</span> <span class="fl">500e6</span>, paired <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>         read_length <span class="op">=</span> <span class="fl">100</span>, sep_files <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>The <span class="math inline">\(F_{ST}\)</span> calculated from the
resulting VCF file could then be compared to output from various
programs to inform which works best in a particular case. For uncertain
population parameters (e.g., migration rates), output from multiple
calls to <code>scrm</code> varying the parameter of interest could be
input to the <code>jackalope</code> pipeline above to identify the
conditions under which one program might have an advantage over
another.</p>
</div>
<div class="section level2">
<h2 id="constructing-a-phylogeny">Constructing a phylogeny<a class="anchor" aria-label="anchor" href="#constructing-a-phylogeny"></a>
</h2>
<div class="section level3">
<h3 id="from-one-phylogenetic-tree">From one phylogenetic tree<a class="anchor" aria-label="anchor" href="#from-one-phylogenetic-tree"></a>
</h3>
<p>This section shows how <code>jackalope</code> can generate haplotypes
from a phylogeny, then simulate sequencing data from those haplotypes to
test phylogeny reconstruction methods. First, I simulated a random
species tree of 10 species, then scaled it to have a maximum tree depth
of <span class="math inline">\(4 N_0\)</span> generations:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tree</span> <span class="op">&lt;-</span> <span class="fu">rcoal</span><span class="op">(</span><span class="fl">10</span><span class="op">)</span></span>
<span><span class="va">tree</span><span class="op">$</span><span class="va">edge.length</span> <span class="op">&lt;-</span> <span class="fl">4</span> <span class="op">*</span> <span class="va">N0</span> <span class="op">*</span> <span class="va">tree</span><span class="op">$</span><span class="va">edge.length</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="fu">node.depth.edgelength</span><span class="op">(</span><span class="va">tree</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Function <code>haps_phylo</code> organizes and checks the
<code>tree</code> object, and including it with the mutation-type
information allowed me to create haplotypes based on this phylogeny:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">haps</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/create_haplotypes.html">create_haplotypes</a></span><span class="op">(</span><span class="va">ref</span>, <span class="fu"><a href="../reference/haps_phylo.html">haps_phylo</a></span><span class="op">(</span><span class="va">tree</span><span class="op">)</span>, <span class="va">sub</span>, <span class="va">ins</span>, <span class="va">del</span><span class="op">)</span></span></code></pre></div>
<p>This results in the following <code>haplotypes</code> object:</p>
<pre><code><span><span class="co">#&gt;                               &lt;&lt; haplotypes object &gt;&gt;</span></span>
<span><span class="co">#&gt; # Haplotypes: 10</span></span>
<span><span class="co">#&gt; # Mutations: 491</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;                           &lt;&lt; Reference genome info: &gt;&gt;</span></span>
<span><span class="co">#&gt; &lt; Set of 4 chromosomes &gt;</span></span>
<span><span class="co">#&gt; # Total size: 985 bp</span></span>
<span><span class="co">#&gt;   name                             chromosome                             length</span></span>
<span><span class="co">#&gt; 2          GAAATCATGTCTTATAAAACGCCAAATC...TCCCCGCTTTCAAGACAGTCTCCTCGAC       252</span></span>
<span><span class="co">#&gt; 3          TCCGTGATCATGCACAAGCATTGCAGTT...TCTAACGGGCGTCGGGCGAATGCTCGCA       240</span></span>
<span><span class="co">#&gt; 4          CACGCGAGCTTACTATTGGTAATGCCTT...CACTTCGGAAGTCAGTTACCAGCTCCTC       253</span></span>
<span><span class="co">#&gt; X          AGTCGATCGCAGACCGTCCTCCTGTAAG...ACACGTCACAGCCACGAAGGCTCACGGT       240</span></span></code></pre>
<p>Now I can generate data for 1 flow cell of <span class="math inline">\(2 \times 250\)</span>bp sequencing on an Illumina
MiSeq v3, where <code>haplotype_barcodes</code> is a character string
that specifies the barcodes for each haplotype. I also wrote the true
phylogenetic tree to a NEWICK file.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">haplotype_barcodes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CTAGCTTG"</span>, <span class="st">"TCGATCCA"</span>, <span class="st">"ATACCAAG"</span>, <span class="st">"GCGTTGGA"</span>,</span>
<span>                        <span class="st">"CTTCACGG"</span>, <span class="st">"TCCTGTAA"</span>, <span class="st">"CCTCGGTA"</span>, <span class="st">"TTCTAACG"</span>, </span>
<span>                        <span class="st">"CGCTCGTG"</span>, <span class="st">"TATCTACA"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/illumina.html">illumina</a></span><span class="op">(</span><span class="va">haps</span>, out_prefix <span class="op">=</span> <span class="st">"phylo_tree"</span>, seq_sys <span class="op">=</span> <span class="st">"MSv3"</span>,</span>
<span>         paired <span class="op">=</span> <span class="cn">TRUE</span>, read_length <span class="op">=</span> <span class="fl">250</span>, n_reads <span class="op">=</span> <span class="fl">50e6</span>,</span>
<span>         barcodes <span class="op">=</span> <span class="va">haplotype_barcodes</span><span class="op">)</span></span>
<span><span class="fu">ape</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/write.tree.html" class="external-link">write.tree</a></span><span class="op">(</span><span class="va">tree</span>, <span class="st">"true.tree"</span><span class="op">)</span></span></code></pre></div>
<p>The true phylogenetic tree would then be compared to the final tree
output from the program(s) the user chooses to test.</p>
</div>
<div class="section level3">
<h3 id="from-gene-trees">From gene trees<a class="anchor" aria-label="anchor" href="#from-gene-trees"></a>
</h3>
<p>Similar to the section above, the ultimate goal here is to test
phylogeny reconstruction methods. The difference in this section is that
instead of using a single, straightforward phylogeny, I use multiple
gene trees per chromosome. In the species used in these simulations,
species 1 diverged from 2 and 3 at <span class="math inline">\(t =
1.0\)</span>, where <span class="math inline">\(t\)</span> indicates
time into the past and is in units of <span class="math inline">\(4
N_0\)</span> generations. Species 2 and 3 diverged at <span class="math inline">\(t = 0.5\)</span>. I assume a recombination rate of
<span class="math inline">\(1 / (4 N_0)\)</span> recombination events
per chromosome per generation. There are 4 diploid individuals sampled
per species. I used the following code to call <code>scrm</code> to
simulate the gene trees and create the object <code>gtrees</code>. And
like the section using segregating sites, we have to wrap
<code>lapply</code> inside <code>list(trees = ...)</code> to replicate
the data structure from <code>scrm</code> for <code>jackalope</code> to
use later.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Run scrm for one chromosome size:</span></span>
<span><span class="va">one_chrom</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">.size</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">sims</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scrm/man/scrm.html" class="external-link">scrm</a></span><span class="op">(</span></span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"24 1"</span>,</span>
<span>              <span class="co"># Output gene trees:</span></span>
<span>              <span class="st">"-T"</span>,</span>
<span>              <span class="co"># Recombination:</span></span>
<span>              <span class="st">"-r 1"</span>, <span class="va">.size</span>,</span>
<span>              <span class="co"># 3 species with no ongoing migration:</span></span>
<span>              <span class="st">"-I 3 8 8 8 0"</span>,</span>
<span>              <span class="co"># Species 2 derived from 1 at time 1.0:</span></span>
<span>              <span class="st">"-ej 1.0 2 1"</span>,  </span>
<span>              <span class="co"># Species 3 derived from 2 at time 0.5:</span></span>
<span>              <span class="st">"-ej 0.5 3 2"</span></span>
<span>        <span class="op">)</span><span class="op">)</span></span>
<span>    <span class="va">trees</span> <span class="op">&lt;-</span> <span class="va">sims</span><span class="op">$</span><span class="va">trees</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span>    <span class="co"># scrm outputs branch lengths in units of 4*N0 generations, but we want just</span></span>
<span>    <span class="co"># generations:</span></span>
<span>    <span class="va">adjust_tree</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">.p</span><span class="op">)</span> <span class="op">{</span></span>
<span>        <span class="co"># Read to phylo object and adjust branch lengths:</span></span>
<span>        <span class="va">.tr</span> <span class="op">&lt;-</span> <span class="fu">read.tree</span><span class="op">(</span>text <span class="op">=</span> <span class="va">.p</span><span class="op">)</span></span>
<span>        <span class="va">.tr</span><span class="op">$</span><span class="va">edge.length</span> <span class="op">&lt;-</span> <span class="va">.tr</span><span class="op">$</span><span class="va">edge.length</span> <span class="op">*</span> <span class="fl">4</span> <span class="op">*</span> <span class="va">N0</span></span>
<span>        <span class="co"># "prefix" from `.p` showing how large the region this gene tree refers to is</span></span>
<span>        <span class="va">prefix</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/strsplit.html" class="external-link">strsplit</a></span><span class="op">(</span><span class="va">.p</span>, <span class="st">"\\]"</span><span class="op">)</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="st">"]"</span><span class="op">)</span></span>
<span>        <span class="co"># Put back together into NEWICK text</span></span>
<span>        <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">prefix</span>, <span class="fu">write.tree</span><span class="op">(</span><span class="va">.tr</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="op">}</span></span>
<span>    <span class="va">trees</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">trees</span>, <span class="va">adjust_tree</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">trees</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="cn">NULL</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="va">trees</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="co"># For all chromosomes:</span></span>
<span><span class="va">gtrees</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>trees <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">ref</span><span class="op">$</span><span class="fu">sizes</span><span class="op">(</span><span class="op">)</span>, <span class="va">one_chrom</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>We can write the true gene trees using <code>write_gtrees</code> to
an <code>ms</code>-style file:</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/haps_gtrees.html">write_gtrees</a></span><span class="op">(</span><span class="fu"><a href="../reference/haps_gtrees.html">haps_gtrees</a></span><span class="op">(</span><span class="va">gtrees</span><span class="op">)</span>, <span class="st">"gtrees"</span><span class="op">)</span></span></code></pre></div>
<p>The <code>create_haplotypes</code> function uses these gene trees to
create variant haplotypes. As for the other haplotype-creation methods,
function <code>haps_gtrees</code> checks and organizes information from
the <code>gtrees</code> object.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">haps</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/create_haplotypes.html">create_haplotypes</a></span><span class="op">(</span><span class="va">ref</span>, <span class="fu"><a href="../reference/haps_gtrees.html">haps_gtrees</a></span><span class="op">(</span><span class="va">gtrees</span><span class="op">)</span>,</span>
<span>                          <span class="va">sub</span>, <span class="va">ins</span>, <span class="va">del</span><span class="op">)</span></span></code></pre></div>
<p>This results in the following <code>haplotypes</code> object:</p>
<pre><code><span><span class="co">#&gt;                               &lt;&lt; haplotypes object &gt;&gt;</span></span>
<span><span class="co">#&gt; # Haplotypes: 24</span></span>
<span><span class="co">#&gt; # Mutations: 2,640</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;                           &lt;&lt; Reference genome info: &gt;&gt;</span></span>
<span><span class="co">#&gt; &lt; Set of 4 chromosomes &gt;</span></span>
<span><span class="co">#&gt; # Total size: 985 bp</span></span>
<span><span class="co">#&gt;   name                             chromosome                             length</span></span>
<span><span class="co">#&gt; 2          GAAATCATGTCTTATAAAACGCCAAATC...TCCCCGCTTTCAAGACAGTCTCCTCGAC       252</span></span>
<span><span class="co">#&gt; 3          TCCGTGATCATGCACAAGCATTGCAGTT...TCTAACGGGCGTCGGGCGAATGCTCGCA       240</span></span>
<span><span class="co">#&gt; 4          CACGCGAGCTTACTATTGGTAATGCCTT...CACTTCGGAAGTCAGTTACCAGCTCCTC       253</span></span>
<span><span class="co">#&gt; X          AGTCGATCGCAGACCGTCCTCCTGTAAG...ACACGTCACAGCCACGAAGGCTCACGGT       240</span></span></code></pre>
<p>To store mutation information by diploid sample, the
<code>write_vcf</code> function writes the <code>haplotypes</code>
object to a VCF file. It assigns every other haplotype to a new diploid
sample using a matrix for the <code>sample_matrix</code> argument:</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/write_vcf.html">write_vcf</a></span><span class="op">(</span><span class="va">haps</span>, out_prefix <span class="op">=</span> <span class="st">"hap_gtrees"</span>,</span>
<span>          sample_matrix <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="va">haps</span><span class="op">$</span><span class="fu">n_haps</span><span class="op">(</span><span class="op">)</span>, ncol <span class="op">=</span> <span class="fl">2</span>, byrow <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Next I generated data for 1 flow cell of <span class="math inline">\(2 \times 250\)</span>bp sequencing on an Illumina
MiSeq v3, with barcodes in the object
<code>haplotype_barcodes</code>.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># 2 of each barcode bc it's diploid</span></span>
<span><span class="va">haplotype_barcodes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"TCGCCTTA"</span>, <span class="st">"CTAGTACG"</span>, <span class="st">"TTCTGCCT"</span>, <span class="st">"GCTCAGGA"</span>, <span class="st">"AGGAGTCC"</span>, </span>
<span>                            <span class="st">"CATGCCTA"</span>, <span class="st">"GTAGAGAG"</span>, <span class="st">"CCTCTCTG"</span>, <span class="st">"AGCGTAGC"</span>, <span class="st">"CAGCCTCG"</span>, </span>
<span>                            <span class="st">"TGCCTCTT"</span>, <span class="st">"TCCTCTAC"</span><span class="op">)</span>, each <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/illumina.html">illumina</a></span><span class="op">(</span><span class="va">haps</span>, out_prefix <span class="op">=</span> <span class="st">"phylo_gtrees"</span>, seq_sys <span class="op">=</span> <span class="st">"MSv3"</span>,</span>
<span>         read_length <span class="op">=</span> <span class="fl">250</span>, n_reads <span class="op">=</span> <span class="fl">50e6</span>, paired <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>         barcodes <span class="op">=</span> <span class="va">haplotype_barcodes</span><span class="op">)</span></span></code></pre></div>
<p>Topologies of the gene trees would then be compared to final
phylogenies output from software the user is interested in testing.
Varying recombination rates or adding gene flow after separation of
species would be natural extensions of these simulations.</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Lucas A. Nell.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
